---
title: "Empirial Cumulative Density Function (ECDF) Analysis on Sample Unknown_B"
author: "Sehrish Kanwal"
date: "`r format(Sys.time(), '%a %Y-%b-%d')`"
output: 
  html_document: 
    toc: true
    toc_depth: 3
    theme: united
    keep_md: yes
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<!-- vim-markdown-toc GFM -->

* [Required R packages](#required-R-packages)
* [Installation](#installation)
* [Result on both bams](#result-on-both-bams)
    * [Console output](#console-output)
    * [Summary of featurecounts command](#summary-of-featurecounts-command)
* [Analysis](#analysis)
    * [Filtering to remove lowly expressed genes](#filtering-to-remove-lowly-expressed-genes)
    * [Distribution Plot](#istribution-plot)
* [Plotting cumulative fraction VS gene log2 fold change](#plotting-cumulative-fraction-vs-gene-log2-fold-change)
    * [Adding Quartiles information and MYC for kallisto bam](#adding-quartiles-information-and-myc-for-kallisto-bam)
    * [Adding Quartiles information and MYC for star bam](#adding-quartiles-information-and-myc-for-star-bam)

<!-- vim-markdown-toc -->

## Required R packages

```{r, message=FALSE}
library(edgeR)
library(limma)
library(Glimma)
library(gplots)
library(ggplot2)
library(dplyr)
library(reshape)
options(width = 140)
library(ggrepel)
```

## Installation

Conda installed subread on spartan. It is available under `/home/sehrishk/.miniconda3/envs/subread/bin/`.

## Result on both bams

The command used to run featurecounts is:

```
/data/cephfs/punim0010/projects/Kanwal_RNASeq_Testing/unknown_B_RNA $ /home/sehrishk/.miniconda3/envs/subread/bin/featureCounts -a /data/cephfs/punim0010/local/development/bcbio/genomes/Hsapiens/GRCh37/rnaseq/ref-transcripts.gtf -o ./featurecount/featurecount_Unknown_B /data/cephfs/punim0010/projects/Kanwal_RNASeq_Testing/unknown_B_RNA/kallisto.bam /data/cephfs/punim0010/projects/Hofmann_WTS/Patients_WTS-merged/final/Unknown_B_RNA/Unknown_B_RNA-ready.bam
```
### Console output

```
//========================== featureCounts setting ===========================\\
||                                                                            ||
||             Input files : 2 BAM files                                      ||
||                           P /data/cephfs/punim0010/projects/Kanwal_RNA ... ||
||                           P /data/cephfs/punim0010/projects/Hofmann_WT ... ||
||                                                                            ||
||             Output file : ./featurecount/featurecount_Unknown_B            ||
||                 Summary : ./featurecount/featurecount_Unknown_B.summary    ||
||              Annotation : /data/cephfs/punim0010/local/development/bcb ... ||
||      Dir for temp files : ./featurecount                                   ||
||                                                                            ||
||                 Threads : 1                                                ||
||                   Level : meta-feature level                               ||
||              Paired-end : no                                               ||
||         Strand specific : no                                               ||
||      Multimapping reads : not counted                                      ||
|| Multi-overlapping reads : not counted                                      ||
||   Min overlapping bases : 1                                                ||
||                                                                            ||
\\===================== http://subread.sourceforge.net/ ======================//

//================================= Running ==================================\\
||                                                                            ||
|| Load annotation file /data/cephfs/punim0010/local/development/bcbio/ge ... ||
||    Features : 1195764                                                      ||
||    Meta-features : 57905                                                   ||
||    Chromosomes/contigs : 61                                                ||
||                                                                            ||
|| Process BAM file /data/cephfs/punim0010/projects/Kanwal_RNASeq_Testing ... ||
||    Paired-end reads are included.                                          ||
||    Assign reads to features...                                             ||
||    Total reads : 108613106                                                 ||
||    Successfully assigned reads : 79414232 (73.1%)                          ||
||    Running time : 4.05 minutes                                             ||
||                                                                            ||
|| Process BAM file /data/cephfs/punim0010/projects/Hofmann_WTS/Patients_ ... ||
||    Paired-end reads are included.                                          ||
||    Assign reads to features...                                             ||
||    Total reads : 109022810                                                 ||
||    Successfully assigned reads : 70065822 (64.3%)                          ||
||    Running time : 4.40 minutes                                             ||
||                                                                            ||
||                         Read assignment finished.                          ||
||                                                                            ||
|| Summary of counting results can be found in file "./featurecount/featurec  ||
|| ount_Unknown_B.summary"                                                    ||
||                                                                            ||
\\===================== http://subread.sourceforge.net/ ======================//
```

The final result `featurecount_Unknown_B` is available on spartan under `/data/cephfs/punim0010/projects/Kanwal_RNASeq_Testing/unknown_B_RNA/featurecount`.

### Summary of featurecounts command

```
Status  kallisto.bam        star.bam
Assigned        79414232        70065822
Unassigned_Unmapped     23066004        6903665
Unassigned_MappingQuality       0       0
Unassigned_Chimera      0       0
Unassigned_FragmentLength       0       0
Unassigned_Duplicate    0       0
Unassigned_MultiMapping 0       11321849
Unassigned_Secondary    0       0
Unassigned_Nonjunction  0       0
Unassigned_NoFeatures   0       14855590
Unassigned_Overlapping_Length   0       0
Unassigned_Ambiguity    6132870 5875884
```

## Analysis

```{r}
seqdata <- read.delim("~/Documents/UMCCR/data/fpkm/featurecount_Unknown_B", stringsAsFactors = FALSE, comment.char = "#")
dim(seqdata)
```

Create a new data object that contains just the counts.

```{r}
countdata <- seqdata[,7:8, drop = FALSE]
head(countdata)
dim(countdata) 
```

Add rownames i.e. GeneIDs to data

```{r}
rownames(countdata) <- seqdata[,1]
head(countdata)

# Taking a look at column names to know the sample names

colnames(countdata)

#Renames sample name to a meaningful title

colnames(countdata) <- c("kallisto.bam", "star.bam")
head(countdata)
```


### Filtering to remove low expressed genes

Genes with very low counts across all libraries provide little evidence for differential expression and they interfere with some of the statistical approximations.

```{r}
myCPM <- cpm(countdata)
head(myCPM)

# Which values in myCPM are greater than 0.15?
thresh <- myCPM > 0.15

# This produces a logical matrix with TRUEs and FALSEs
head(thresh)

# Summary of how many TRUEs there are in each row
# There are 18128 genes (out of 57905) that have TRUEs in both bams.
table(rowSums(thresh))

# Would like to keep genes that have TRUE thresh value for both bams in each row of thresh
keep <- rowSums(thresh) == 2

# Subset the rows of countdata to keep the more highly expressed genes
counts_keep <- countdata[keep, ,  drop = FALSE]

# Taking a look at the subset data
summary(keep)
head(counts_keep)
dim(counts_keep)
str(counts_keep)
```

A CPM of 0.15 is used as it corresponds to a count of 10-15 for the library sizes in this data set. If the count is any smaller, it is considered to be very low, indicating that the associated gene is not expressed in that sample. As a general rule, a good threshold can be chosen by identifying the CPM that corresponds to a count of 10, which in this case is about 0.15. It is important to filter with CPMs rather than filtering on the counts directly, as the latter does not account for differences in library sizes between samples.

```{r}
# Let's have a look and see whether threshold of 0.25 does indeed correspond to a count of about 10-15 in 1st sample (star bam)

plot(myCPM[,1],countdata[,1])

# Limit the x and y-axis so we can actually look to see what is happening at the smaller counts

plot(myCPM[,1],countdata[,1],ylim=c(0,50),xlim=c(0,3))

# Add a vertical line at 0.25 CPM and horizontal at ~10
abline(v = 0.15, h = 10, col=c("blue", "red"))

# Let's have a look and see whether threshold of 0.25 does indeed correspond to a count of about 10-15 in 2nd sample (kallisto bam)

plot(myCPM[,2],countdata[,2])

# Limit the x and y-axis so we can actually look to see what is happening at the smaller counts

plot(myCPM[,2],countdata[,2],ylim=c(0,50),xlim=c(0,2))

# Add a vertical line at 0.25 CPM and horizontal at ~10
abline(v = 0.15, h = 10, col=c("blue", "red"))

```

### Distribution Plot

Count data is not normally distributed, so if we want to examine the distributions of the raw counts we need to log the counts. Next weâ€™ll use box plots to check the distribution of the read counts on the log2 scale. We can use the cpm function to get log2 counts per million, which are corrected for the different library sizes. The cpm function also adds a small offset to avoid taking log of zero.

```{r}
# First, create a DGEList object. This is an object used by edgeR to store count data. It has a number of slots for storing various parameters about the data.
y <- DGEList(counts_keep)

# have a look at y
y

# See what slots are stored in y
names(y)

# Library size information is stored in the samples slot
y$samples

# Get log2 counts per million
logcounts <- cpm(y,log=TRUE)

# have a look at logcounts
head(logcounts)
dim(logcounts)

# Check distributions of samples using boxplots
boxplot(logcounts, xlab="", ylab="Log2 counts per million",las=2)

# Let's add a blue horizontal line that corresponds to the median logCPM
abline(h=median(logcounts),col="blue")
title("Boxplots of logCPMs (unnormalised)")
```

## Plotting cumulative fraction VS gene log2 count values

For the fold change, I'll need to have a comparison e-g. against a normal sample (which I am not doing currently). According to Olivers's comment on this "We won't be running a 'normal' RNA sample for the time being. That means following the POG / Illumina approach of using unrelated normals as reference -- we'll need to build up a small tissue atlas using TCGA (or selected in-house samples) where we have RNA from matching normal tissue. Log-fold changes are then relative to the overall reference data, or the subset we think best fits the bill. Count data will do for now until we get that data set."

For this task, Empirical Cumulative Density Function (ecdf) is used.

```{r}
df <- as.data.frame(logcounts)

# Plot stat_ecdf for both kallisto and star bam on the same plot
dff <- data.frame(x = c(df$kallisto.bam, df$star.bam), g = gl(2, 21514, labels = c("kallisto", "star")))

p <- ggplot(dff, aes(x, colour = g)) + 
  stat_ecdf(geom = "step") +
  labs(title="Empirical Cumulative Density Function", y = "cumulative fraction", x="gene log2 count value in bams") +
  geom_vline(xintercept = 0, linetype="dotted", color = "blue")

p <- p + expand_limits(x = -5, y = 0)
p
```

### Adding Quartiles information and MYC for kallisto bam

It can be clearly observed that MYC is highly expressed (log 2 count value = 9.345).

```{r}
df <- as.data.frame(logcounts)

#Preparing quartiles
q1 <- quantile(df$kallisto.bam)[2]
q2 <- quantile(df$kallisto.bam)[3]
q3 <- quantile(df$kallisto.bam)[4]

#Making a tibble from quartiles
events <- data_frame(quartiles = c(q1,q2,q3),
                     text = c('Q1','Q2','Q3'))

# Plot stat_ecdf for kallisto 
p <- ggplot(df, aes(df$kallisto.bam)) + 
  stat_ecdf(geom = "step", size = 0.25) +
  geom_vline(data = events, aes(xintercept = events$quartiles), color = "red") +
  geom_text(data = events, mapping = aes(label = events$text, y = 0, x = quartiles), family="Times", size = 3, hjust = 0) +
  geom_point(aes(x=df["ENSG00000136997", 1], y=0.99 , colour = "yellow"), show.legend = FALSE, label="MYC") +
  geom_text(aes(x=9.4, label="MYC", y=0.96), family="Times", size = 3) +
  labs(title="Empirical Cumulative Density Function", y = "cumulative fraction", x="gene log2 count change in kallisto bam")

p <- p + expand_limits(x = -5, y = 0)
p
```

### Adding Quartiles information and MYC for star bam

It can be clearly observed that MYC is highly expressed (log 2 count value = 9.484).

```{r}
df <- as.data.frame(logcounts)

#Preparing quartiles
q1 <- quantile(df$star.bam)[2]
q2 <- quantile(df$star.bam)[3]
q3 <- quantile(df$star.bam)[4]

#Making a tibble from quartiles
events <- data_frame(quartiles = c(q1,q2,q3),
                     text = c('Q1','Q2','Q3'))

# Plot stat_ecdf for kallisto 
p <- ggplot(df, aes(df$star.bam)) + 
  stat_ecdf(geom = "step", size = 0.25) +
  geom_vline(data = events, aes(xintercept = events$quartiles), color = "red") +
  geom_text(data = events, mapping = aes(label = events$text, y = 0, x = quartiles), family="Times", size = 3, hjust = 0) +
  geom_point(aes(x=df["ENSG00000136997", 2], y=0.99 , colour = "yellow"), show.legend = FALSE, label="MYC") +
  geom_text(aes(x=9.4, label="MYC", y=0.96), family="Times", size = 3) +
  labs(title="Empirical Cumulative Density Function", y = "cumulative fraction", x="gene log2 count change in star bam")

p <- p + expand_limits(x = -5, y = 0)
p
```



