# Read count data analysis

This repository documents pipeline for analysing RNA-seq read count data.


## Table of contents

<!-- vim-markdown-toc GFM -->
* [Workflow](#workflow)
* [Merge per-sample read count data](#merge-per-sample-read-count-data)
* [Combine data from different datasets](#combine-data-from-different-datasets)
* [Batch effects assessment](#batch-effects-assessment)
* [Differential expression analysis](#differential-expression-analysis)

<!-- vim-markdown-toc -->
<br>

## Workflow

Overall RNA-seq data analysis workflow using pre-processed pre-sample read count data. The analysis can start from *per-sample read count files*, similar to those generated by [bcbio RNA-seq pipeline](https://bcbio-nextgen.readthedocs.io/en/latest/contents/pipelines.html#rna-seq), *read conunt data matrix* or *combined/normalised expression matrix*.

![Read count data analysis workflow](../../img/RNA-seq_data_analysis_workflow.jpeg?raw=true "Read count data analysis workflow")


## Merge per-sample read count data

To merge multiple read count files use the *[mergeCounts2Matrix.R](https://github.com/umccr/RNA-seq-analysis/blob/z_score/bam/featurecount/mergeCounts2Matrix.R)* script. Each per-sample read count file should have two columns containing gene and read count information without a header (see this [example file](TCGA-PAAD/TCGA-2J-AAB1-01A-11R-A41B-07.htseq.counts)). All read count files should be located in one directory. One also needs to create so called **target file** with four columns:

1. "Sample_name" - this lists prefered sample names that will appear in the merged matrix and any tables or plots in downstream analyses
2. "File_name" - name of the corresponding read count file
3. "Target" - biological groupb/phenotypes that will be used to annotate samples and to perform downstream analyses, including differential expression analysis
4. "Replicates" - indication of any technical replicates to take into account in differential expression analysis 


To facilitate analyses and output files organisation the target file is expected to be in user-defined **project directory**. This is where all the resultant files will be saved.


### Arguments

**Script**: *[mergeCounts2Matrix.R](https://github.com/umccr/RNA-seq-analysis/blob/z_score/bam/featurecount/mergeCounts2Matrix.R)*

Argument | Description
------------ | ------------
--projectDir | Project directory. This is where the target file is expected and where the merged matrix will be saved
--target | Name of the target file. It expects to have four columns: (1) Sample_name, (2) File_name, (3) Target and (4) Replicates
--inDir | Directory containing per-sample expression files. Note that only files listed in the target file will be used to generate the merged matrix. No header is expected. The sample names in the merged matrix will be added based on the sample names in the target file
--outFile | Core name for the merged matrix output file, to which ".counts.matrix.txt" suffix wil be added 
<br />

**Command line use example**:

```
Rscript mergeCounts2Matrix.R --projectDir /Combined_data --target /TCGA_PAAD_Target.txt --inDir /TCGA-PAAD --outFile TCGA-PAAD
```
<br>

### Output files

This will read read count files listed in *TCGA_PAAD_Target.txt* target file and located in the */TCGA-PAAD* folder, and will generate the following output files in the *Combined_data* project directory:

Output file | Description
------------ | -----------
TCGA-PAAD.counts.matrix.txt | Merged matrix with read counts for individual genes (rows) across all per-sample (columns)
TCGA-PAAD.mergeCounts2Matrix.parameters.txt | File reporting used parameters
TCGA-PAAD.mergeCounts2Matrix.missing_files.txt | File created in case there are read count files listed in the *target file* but absent in the folder with per-sample expression files
TCGA-PAAD.mergeCounts2Matrix.missing_genes.txt | File created in case there are genes that were not present across all per-sample expression files and thus were not included in the *merged expression matrix*
<br />

The output files will be organised following the folder structure below

```
Combined_data
|
|____TCGA_PAAD_Target.txt
|____TCGA-PAAD.counts.matrix
|____mergeCounts2Matrix.parameters.txt
|____TCGA-PAAD.mergeCounts2Matrix.missing_files.txt (optional)
|____TCGA-PAAD.mergeCounts2Matrix.missing_genes.txt (optional)

```
<br />



## Combine data from different datasets

<br>

## Batch effects assessment

<br>

## Differential expression analysis

<br>
