# RNA-seq-analysis

This repository documents scripts for combining RNA-seq count data and exploring data distribution patters of transformed data as whole as well as for the genes of interest.

## Table of contents

<!-- vim-markdown-toc GFM -->
* [Read count data combination](#read-count-data-combination)
	* [Workflow](#workflow)
	* [Merge per-sample read count data](#merge-per-sample-read-count-data)
	* [Combine data from different datasets](#combine-data-from-different-datasets)
* [Data distribution evaluation](#data-distribution-evaluation)

<!-- vim-markdown-toc -->
<br>

## Read count data combination

This section describes steps and scripts used to combine RNA-seq read count data from multiple samples and datasets.

### Workflow

The figure below is an overall workflow combining pre-processed and per-sample RNA-seq read count data. The analysis can start from *per-sample read count files*, similar to those generated by [bcbio RNA-seq pipeline](https://bcbio-nextgen.readthedocs.io/en/latest/contents/pipelines.html#rna-seq), *read count data matrix* or *combined/normalised expression matrix*. The data combination includes filtering out genes with low counts (optional), transformation, log-transformation (optional) and normalisation.


![Read count data analysis workflow](img/RNA-seq_analysis_workflow.png?raw=true "Read count data analysis workflow")

## Merge per-sample read count data

To merge multiple read count files use the *[mergeCounts2Matrix.R](./readcount-analysis/mergeCounts2Matrix.R)* script. Each per-sample read count file should have two columns containing gene and read count information without a header (see this [example read count file](./readcount-analysis/TCGA-PAAD/TCGA-2J-AAB1-01A-11R-A41B-07.htseq.counts)). Read count files for all samples should be located in one directory. Also, the user needs to create a tab-delimited **target file** with four columns (see this [example target file](./readcount-analysis/TCGA-PAAD/TCGA_PAAD_Target.txt)):

1. `Sample_name` - this coloumn specifies preferred sample names that will appear in the merged matrix and any tables or plots in downstream analyses
2. `File_name` - name of the corresponding read count file
3. `Target` - biological group/phenotypes that will be used to annotate samples and to perform downstream analyses
4. `Replicates` - indication of any technical replicates to take into account, i.e. for differential expression analysis (can be left blank)

To facilitate downstream analyses and files organisation, the output files will be saved in user-defined **project directory** with user-defined prefix (core name).

### Arguments

**Script**: *[mergeCounts2Matrix.R](./readcount-analysis/mergeCounts2Matrix.R)*

Argument | Description
------------ | ------------
--projectDir | Project directory. This is where the merged matrix will be saved
--target | Name and location of the target file. The target file is expected to have four columns: (1) Sample\_name, (2) File\_name, (3) Target and (4) Replicates
--inDir | Directory containing per-sample expression files. Note that only files listed in the target file will be used to generate the merged matrix. No header is expected. The sample names in the merged matrix will be added based on the sample names in the target file
--outFile | Core name for the merged matrix output file, to which ".counts.matrix.txt" suffix will be added 
<br/>

**Command line use example**:

```
data=/data/cephfs/punim0010/projects/Jacek_RNA_seq_workflows/RNA-seq-analysis/readcount-analysis

Rscript mergeCounts2Matrix.R --projectDir $data/Combined_data --target $data/TCGA-PAAD/TCGA_PAAD_Target.txt --inDir $data/TCGA-PAAD --outFile TCGA-PAAD
```

<br><br>

The input and output files will be organised following the folder structure below

```
|
|____Project_directory (User-defined)
| |____[core_name].counts.matrix
| |____mergeCounts2Matrix.parameters.txt
| |____[core_name].mergeCounts2Matrix.missing_files.txt (optional)
| |____[core_name].mergeCounts2Matrix.missing_genes.txt (optional)
|
|____Count_data_folder
  |____Count_file_sample_1
  |____Count_file_sample_2
  ...
  |____Count_file_sample_n
  |____Target_file
```

<br/>

### Output files

[The example command above](#arguments) will read count files listed in *TCGA_PAAD_Target.txt* target file and located in the *[data]/TCGA-PAAD* folder, and will generate the following output files in the *[data]/Combined_data* project directory:

Output file | Description
------------ | -----------
TCGA-PAAD.counts.matrix.txt | Merged matrix with read counts for individual genes (rows) across all samples (columns)
TCGA-PAAD.mergeCounts2Matrix.parameters.txt | File reporting used parameters
TCGA-PAAD.mergeCounts2Matrix.missing_files.txt | File created in case there are read count files listed in the *target file* but absent in the folder with per-sample expression files
TCGA-PAAD.mergeCounts2Matrix.missing_genes.txt | File created in case there are genes that were not present across all per-sample expression files and thus were not included in the *merged expression matrix*
<br/>


## Combine data from different datasets

To combine gene-by-sample read count matrices from different datasets use the *[combineExprData.R](./readcount-analysis/combineExprData.R)* script. It collects user-defined parameters and pass them to *[combineExprData.Rmd](./readcount-analysis/combineExprData.Rmd)*, which generates combined data matrix, accompanying html report, files and plots. Note, only **genes intersection** across all datasets expression matrices will be reported in the combined expression matrix. The pipeline is based on recommendaitons from *[RNAseq123](https://master.bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html)* package.

*[combineExprData.R](./readcount-analysis/combineExprData.R)* script requires manually prepared tab-delimited **datasets file** with four columns (*Dataset_name*, *Expression_matrix*, *Target_file* and *Outliers_file*) to define names of the datasets to be merged, the correspoding expression matrices and target files, samples's names for the merged matrix, as well as the files listing outlier samples to be removed before combining the data (see this [example datasets  file](./readcount-analysis/Combined_data/Datasets_list.txt)):

1. `Dataset_name` - this coloumn specifies preferred dataset name that will appear in any tables or plots in downstream analyses
2. `Expression_matrix` - location and name of the corresponding read count matrix
3. `Target_file` - location and name of the corresponding target file
4. `Outliers_file` - location and name of the corresponding outliers file listing outlier samples to be removed before combining the data (this can be left blank)

An example of **target file** is [here](./readcount-analysis/TCGA-PAAD/TCGA_PAAD_Target.txt) and is described in [Merge per-sample read count data](#merge-per-sample-read-count-data) section in this repo. An example of **outliers file** is [here](./readcount-analysis/Combined_data/TCGA-PAAD_outliers.txt). This file, however, is not required if no samples in the corresponding dataset are to be removed.

To facilitate analyses and output files organisation, the **datasets file** is expected to be in user-defined **project directory**. This is where all the resultant files will be saved. The input and ouput directories structure are as follows:

```
|
|____Project_directory (User-defined)
| |____Outliers_file_dataset_1 (optional)
| |____Outliers_file_dataset_2 (optional)
|
|____Count_data_folder_dataset_1
| |____Count_matrix_dataset_1
| |____Target_file_dataset_1
|
|____Count_data_folder_dataset_2
  |____Count_matrix_dataset_2
  |____Target_file_dataset_2
```

### Arguments

**Script**: *[combineExprData.R](./readcount-analysis/combineExprData.R)*

Argument | Description
------------ | ------------
--projectDir | Project directory. This is where the datasets file is expected and where the merged matrix will be saved
--datasets | Name of the datasets file listing info about datasets to combine 
<br/>

**Command-line use example**:

```
Rscript combineExprData.R --projectDir $data/Combined_data --datasets Datasets_list.txt
```

<br><br>

The input and output files will be organised following the folder structure below

```
|
|____Project_directory (User-defined)
| |____Outliers_file_dataset_1 (optional)
| |____Outliers_file_dataset_2 (optional)
| |____[datasets_names]_TMM.txt
| |____[datasets_names].combineExprData.parameters.txt
| |____[datasets_names].missing_genes.txt (optional)
| |____[datasets_names]_RNAseq_libSize_datasets.html
| |____[datasets_names]_RNAseq_libSize_targets.html
| |____[datasets_names]_filtering.pdf
| |____[datasets_names]_normalisation.pdf
| |____[datasets_names]_scaling_factor_datasets.html
| |____[datasets_names]_scaling_factor_targets.html
| |____[datasets_names]_plots
|   |____[sample_1]_MDplot.pdf
|   |____[sample_2]_MDplot.pdf
|   ...
|   |____[sample_n]_MDplot.pdf
|
|____Count_data_folder_dataset_1
| |____Count_matrix_dataset_1
| |____Target_file_dataset_d1
|
|____Count_data_folder_dataset_2
  |____Count_matrix_dataset_2
  |____Target_file_dataset_2
```

<br/>

### Output files

[The example command above](#arguments-1) will read *count matrices*, *target* and *outliers* files listed in *Datasets_list.txt* datasets file located within the *[data]/Combined_data* project folder. Based on the information about the input files the *[combineExprData.Rmd](./readcount-analysis/combineExprData.Rmd)* script will generate **[Datasets_list.txt.combineExprData.html](./readcount-analysis/Combined_data/Datasets_list.txt.combineExprData.html) report** with interactive summary plots and the following output files in the *[data]/Combined_data* project directory:

Output file | Description
------------ | -----------
[Avner_TCGA-PAAD_TMM.txt](./readcount-analysis/Combined_data/Avner_TCGA-PAAD_TMM.txt) | Combined and normalised [TMM](https://www.ncbi.nlm.nih.gov/pubmed/20196867) matrix with *log-cpm* values for individual genes (rows) across all samples (columns)
[Avner_TCGA-PAAD.combineExprData.parameters.txt](./readcount-analysis/Combined_data/Avner_TCGA-PAAD.combineExprData.parameters.txt) | File reporting used parameters
[Avner_TCGA-PAAD.missing_genes.txt](./readcount-analysis/Combined_data/Avner_TCGA-PAAD.missing_genes.txt) | File created in case there are genes that were not present across all per-dataset count matrices and thus were not included in the *combined expression matrix*
[Avner_TCGA-PAAD_RNAseq_libSize_datasets.html](./readcount-analysis/Combined_data/Avner_TCGA-PAAD_RNAseq_libSize_datasets.html) | Interactive bar-plot illustrating library size for each sample (bar). The colours indicate datasets
[Avner_TCGA-PAAD_RNAseq_libSize_targets.html](./readcount-analysis/Combined_data/Avner_TCGA-PAAD_RNAseq_libSize_targets.html) | Interactive bar-plot illustrating library size for each sample (bar). The colours indicate sample groups, as provided in the target files.
[Avner_TCGA-PAAD_filtering.pdf](./readcount-analysis/Combined_data/Avner_TCGA-PAAD_filtering.pdf) | Histograms with data distribution before and after low expressed genes filtering
[Avner_TCGA-PAAD_normalisation.pdf](./readcount-analysis/Combined_data/Avner_TCGA-PAAD_normalisation.pdf) | Box-plots of *log-cpm* for individual samples, coloured by dataset, before and after [TMM](https://www.ncbi.nlm.nih.gov/pubmed/20196867) normalisation
[Avner_TCGA-PAAD_MD_plots](./readcount-analysis/Combined_data/Avner_TCGA-PAAD_MD_plots) | Folder with [mean-difference (MD) plots](./readcount-analysis/Combined_data/Avner_TCGA-PAAD_MD_plots/TCGA.2J.AAB4.01A.12R.A41B.07_MDplot.pdf) for all samples demonstrating the *library size-adjusted log-fold change* between two libraries (the difference) against the *average log-expression* across those libraries (the mean)
[Avner_TCGA-PAAD_scaling_factor_datasets.html](./readcount-analysis/Combined_data/Avner_TCGA-PAAD_scaling_factor_datasets.html) | Interactive scatter-plot with library sizes of individual samples plotted against corresponsing scaling factors. The colours indicate datasets
[Avner_TCGA-PAAD_scaling_factor_targets.html](./readcount-analysis/Combined_data/Avner_TCGA-PAAD_scaling_factor_targets.html) | Interactive scatter-plot with library sizes of individual samples plotted against corresponsing scaling factors. The colours indicate sample groups, as provided in the target files

<br/>

###### Note

Read count data combined from various resources, datasets or batches will be confounded by non-biological experimental variances that prevent direct comparison of samples from different studies. These inevitable variations may be due to differences in sample preparation or processing protocols, as well as time of the day when the assays were performed. The discrepancies observed across different experiments may influence the consistency and reliability of downstream analysis, so it is important to check for possible batch effects and adjust expression measurements to minimise the variance caused by confounding factors. Currently, however, this part of the workflow is not implemented yet.

<br>

## Data distribution evaluation

Use [combinedExprDataDistribution.R](./readcount-analysis/combinedExprDataDistribution.R) script to investigate the expression data distribution patters of user-defined genes (max 10) in [combined expression data](#read-count-data-combination) derived from different samples. The script generates a report with interactive plots illustrating changes in the data introduced at individual processing steps and facilitate catching pontetial problems due to applied proecssing methods.

The script allows to process the read count data using combonation of the following transformation and normalisation methods:


Transformation | TMM | TMMwzp | RLE | Upperquartile | Quantile | None
------------ | ------------ | ------------ | ------------ | ------------ | ------------ | ------------
CPM | Yes | Yes | Yes | Yes | - | Yes
TPM | - | - | - | - | Yes | Yes
<br/>

**CPM** (Counts Per Million) are obtained by dividing counts by the library counts sum and multiplying the results by a million.

**TPM** (Transcripts Per Million) are obtained by dividing read counts by gene lengths (RPK, Reads Per Kilobases). These RPK are then divided by the sum of all RPK values in a sample and multiplying the results by 1 million.

CPM is depth-normalised counts whereas TPM is length normalised (and then normalized by the length-normalized values of the other genes).

*[combinedExprDataDistribution.R](./readcount-analysis/combinedExprDataDistribution.R)* script collects user-defined parameters and pass them to *[combinedExprDataDistribution.Rmd](./readcount-analysis/combinedExprDataDistribution.Rmd)*, which generates html report with interactive plots and tables.

### Arguments

**Script**: *[combinedExprDataDistribution.R](./readcount-analysis/combinedExprDataDistribution.R)*

Argument | Description
------------ | ------------
--exprDir | Directory with expression data. This is where the combined expression matrix and accompanying files will be saved
--exprFile | File with expression data (read counts)
--annotFile | Samples annotation file with four columns: (1) *Sample_name*, (2) *File_name* (may be balnk), (3) *Target* and (4) *Replicates* (may be balnk)
--transform | Transformation method to be used when converting read counts. Available options are: *CPM* (defualt) and *TPM*
--norm | Normalisation method. *TMM*, *TMMwzp*, *RLE* and *upperquartile* methods are available for *CPM-transformed* data and *quantile* normalisation is used for *TPM-transformed* data. *None* (default) is available for both transformation methods
--filter | Filtering out low expressed genes. Available options are: *TRUE* (defualt) and *FALSE*
--log | Log (base 2) transform data before normalisation. Available options are: *TRUE* (defualt) and *FALSE*
--genes | List of genes to be considered. Up to 10 genes are allowed, each separated by comma
--ensembl | Is input data annotated using ensembl gene IDs? Available options are: *TRUE* (defualt) and *FALSE*
--samples | ID of samples of interest (OPTIONAL)
--results_name | Desired core name for the results folder
<br/>

**Command-line use example**:

```
Rscript combinedExprDataDistribution.R --exprDir $data/Combined_data --exprFile UMCCR_PC.counts.matrix.txt.subset.txt --annotFile UMCCR_PC_Target_cleaned.txt --transform CPM --norm TMM --filter TRUE --log TRUE --genes KRAS,CDKN2A,TP53,SMAD4 --samples CCR170012_MH17T001P013,CCR180029_MH18T002P038_RNA --results_name expression_distributions_results 
```

<br>

### Output report

The example command above will generate ***[expression_distributions_results_CPM_TMM.html report](./readcount-analysis/Combined_data/expression_distributions_results_CPM_TMM.html)*** with interactive summary plots and the following output files in the *[data]/Combined_data* project directory:

1. Bar-plot illustrating library size for each sample
2. Histograms presenting distributions of transformed data before and after data filtering (optional step)
3. Box-plots presenting transformed (and filtered) data for individual samples, coloured by sample groups, before and after normalisation procedure (optional step)
4. Box-plot illustrating z-score transformed data for individual samples, coloured according to corresponding groups
5. Scatter-plots illustrating the effect of data normalisation (optional step) and z-score transformation on the transformed data for user-defined genes, as well as per-sample normlaisation factors indicated by a colour shading
6. Plots presenting transformed, normalised (optional step) and z-score data distribution for selected gene(s) along with simulated normal, binomial (p=0.25 and p=0.75) and bimodal distributions (computed based on the input data)

<br>
