---
title: "Investigate expression data for selected genes"
author: "UMCCR"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
  exprDir: 
  exprFile: 
  annotFile: 
  transform: 
  norm: 
  filter: 
  log: 
  genes: 
  ensembl: 
  samples: 
  results_name: 
  scaling: 
  ensembl_version: 86
output:
  html_document:
    keep_md: yes
    code_folding: hide
#    toc: true
#    toc_float: true
---

The aim of this report is investigate the expression data distribution patterns of user-defined genes (max 10) in combined expression data (**`r params$exprFile`**) derived from different samples. The data combination includes filtering out genes with low counts `r if ( !params$filter ) { c("(this step is skipped)") }`, transformation (`r params$transform` method, `r if ( !params$log ) { c("with no log-transformation") } else if ( params$log ) { c("followed by log2-transformation") }`), normalisation (`r params$norm`) and scaling (`r if ( tolower(params$scaling) == "sample-wise" ) { c("Z-score tansformation") } else { c("centering") }`). The pipeline is based on recommendations from *[RNAseq123](https://master.bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html){target="_blank"}* package.

```{r define_functions, comment=NA, message=FALSE, warning=FALSE}
### Define functions

##### Create 'not in' operator
"%!in%" <- function(x,table) match(x,table, nomatch = 0) == 0
##### Prepare object to write into a file
prepare2write <- function (x) {
  
  x2write <- cbind(rownames(x), x)
  colnames(x2write) <- c("Gene",colnames(x))
  return(x2write)
}
##### Prepare gene data matrix to write into a file
geneMatrix2write <- function (x) {
  
  x2write <- cbind(rownames(x), x)
  colnames(x2write) <- c("Gene",colnames(x))
  return(x2write)
}

##### Assign colours to different groups
getTargetsColours <- function(targets) {
  
  ##### Predefined selection of colours for groups
  targets.colours <- c("red","blue","green","darkgoldenrod","darkred","deepskyblue", "coral", "cornflowerblue", "chartreuse4", "bisque4", "chocolate3", "cadetblue3", "darkslategrey", "lightgoldenrod4", "mediumpurple4", "orangered3","indianred1","blueviolet","darkolivegreen4","darkgoldenrod4","firebrick3","deepskyblue4", "coral3", "dodgerblue1", "chartreuse3", "bisque3", "chocolate4", "cadetblue", "darkslategray4", "lightgoldenrod3", "mediumpurple3", "orangered1")
  
  f.targets <- factor(targets)
  vec.targets <- targets.colours[1:length(levels(f.targets))]
  targets.colour <- rep(0,length(f.targets))
  for(i in 1:length(f.targets))
    targets.colour[i] <- vec.targets[ f.targets[i]==levels(f.targets)]
  
  return( list(vec.targets, targets.colour) )
}

##### Assign colours to different distributions
getGenesColours <- function(genes) {
  
  ##### Predefined selection of colours for genes
  genes.colours <- rainbow(length(genes))

  f.genes <- factor(genes)
  vec.genes <- genes.colours[1:length(levels(f.genes))]
  genes.colour <- rep(0,length(f.genes))
  for(i in 1:length(f.genes)){
    genes.colour[i] <- vec.genes[ f.genes[i]==levels(f.genes)]
  }
  
  return( genes.colour )
}

##### Assign colours to different samples
getSamplesColours <- function(samples) {
  
  ##### Predefined selection of colours for genes
  samples.colours <- brewer.pal(length(samples), "Set1")

  f.samples <- factor(samples)
  vec.samples <- samples.colours[1:length(levels(f.samples))]
  samples.colour <- rep(0,length(f.samples))
  for(i in 1:length(f.samples)){
   samples.colour[i] <- vec.samples[ f.samples[i]==levels(f.samples)] 
  }
  
  return( samples.colour )
}

##### Calculate TPM from RPKM (from http://luisvalesilva.com/datasimple/rna-seq_units.html )
tpm_from_rpkm <- function(x) {
  rpkm.sum <- colSums(x)
  return(t(t(x) / (1e-06 * rpkm.sum)))
}

##### Convert density to counts
density2freq <- function(density) {

  freq = length(density)/sum(density) * density
  return(freq)
}
  

##### Generate scatter plots for samples (highlighting the samples and genes of interest) to illustrate the effect of data normalization and z-score transformation/centering. Currently, this function works for one gene at a time.
comparePlot <- function(genes, dataset1, dataset2, main_title, x_title, y_title, sampleName, plot_mode = "static") {
  
  ##### Used data for user-defined genes
  dataset1 <- dataset1[ genes, ,drop=FALSE]
  dataset2 <- dataset2[ genes, ,drop=FALSE]
 
  ##### Make sure that samples are in the same order
  dataset2 <- dataset2[ , colnames(dataset1) ,drop=FALSE]
  
  #### Interactive genes distribution plot
  ##### Create empty data frame
  data.df <- data.frame(matrix(ncol = 4, nrow = 0))
  colnames(data.df) <- c("gene", "sample", "dataset1", "dataset2")
  
  for (i in 1:nrow(dataset1)) {
      data.df <- rbind(data.df, data.frame( gene = genes[i], sample = names(sort(dataset1[ genes[i], ])), dataset1 = dataset1[ genes[i], ], dataset2 = dataset2[ genes[i], ]))
  }
  
  ##### Extract expression for selected sample in the distributions dataframe
  data.df.selected <- data.frame(matrix(ncol = 5, nrow = 0))
  colnames(data.df.selected) <- c("data", "gene", "sample", "expr", "dens")
  
  for ( i in 1:length(sampleName) ) {
    data.df.selected <- rbind(data.df.selected, data.df[sampleName[i] == data.df$sample, ])
  }
  
  ##### Assign colours to distributions
  genes.colour <- getGenesColours(unique(data.df$gene))
  
  ##### Assign colours to samples
  sample.colour <- getSamplesColours(unique(data.df.selected$sample))

  
  ##### Generate interactive density plot
  p <- plotly::plot_ly(data.df, x = ~dataset1, y = ~dataset2, type = 'scatter', color = ~gene, colors = genes.colour, text = data.df$sample, width = 00, height = 400) %>%
    plotly::add_markers(y = data.df.selected$dataset2, x = data.df.selected$dataset1, 
                name = data.df.selected$sample,
                text = data.df.selected$sample,
                mode = 'markers',
                marker = list(size = 8, colors = data.df.selected$sample, color = rep(sample.colour, each = nrow(data.df.selected)/length(samples)), line = list(color = "grey", width = 2)),
                showlegend = TRUE, inherit = FALSE) %>%
    
    ##### Add Loess smoothed line
    plotly::add_lines(x = data.df$dataset1, y = ~fitted(loess(data.df$dataset2 ~ data.df$dataset1)),
            line = list(color = 'grey'),
            name = "Loess Smoother", opacity = 0.5, showlegend = TRUE) %>%
    plotly::layout(title = main_title,
           xaxis = list(title = x_title),
           yaxis = list (title = y_title))
  
return( p )
}

##### Generate density and expression distribution plots for selected genes, highlighting samples of interest. Currently, this function works for one gene at a time.
densityPlot <- function(genes, data, main_title, x_title, sampleName, distributions = NULL, plot_mode = "static") {
  
  ##### Used data for user-defined genes
  data <- data[ genes, ,drop=FALSE]
  
  ##### Transformed data for html report
  #### Interactive genes distribution plot
  ##### Create empty data frame
  data.df <- data.frame(matrix(ncol = 4, nrow = 0))
  colnames(data.df) <- c("gene", "sample", "expr", "dens")
  
  for (i in 1:nrow(data)) {
      data.df <- rbind(data.df, data.frame(gene = genes[i], sample = names(sort(data[ genes[i], ])), expr = sort(data[ genes[i], ]), dens = density2freq(density(data[i,], n=ncol(data))$y)))
  }
  
  ##### Generate values to generate various distributions
  if ( !is.null(distributions) && length(genes) == 1) {
    
    ##### Use the density values obtained from the expression values
    expr.sorted <- sort(data[ genes, ])
    
    ##### Get min and max values based on the expression data
    data.x <- seq(min(expr.sorted), max(expr.sorted), length.out = ncol(data))
    
    ##### Create empty data frame
    data.df.dist <- data.frame(matrix(ncol = 4, nrow = 0))
    colnames(data.df.dist) <- c("gene", "sample", "expr", "dens")
    
    ##### Generate y-values to mirror distributions of interest
    ##### Generate y-values for normal distribution. Useful resource https://stats.idre.ucla.edu/r/modules/probabilities-and-distributions/
    if ( "normal" %in% tolower(distributions)) {
      
      ##### Generate y-values for normal distribution
      data.y <- dnorm(data.x, mean = mean(data.x), sd = (max(data.x)-mean(data.x))/5)
      data.df.dist <- rbind(data.df.dist, data.frame(gene="Normal distribution", sample = names(sort(data[ genes, ])), expr=data.x, dens=density2freq(data.y)))
    } 
    
    if ( "binomial" %in% tolower(distributions) ) {
      
      ##### Generate x- and y-values for binomial distribution. Useful link https://stat.ethz.ch/R-manual/R-devel/library/stats/html/Binomial.html
      #data.x <- round(seq(min(data), max(data), length.out = ncol(data)), digits = 0)
      data.x <- 1:ncol(data)
      data.y <- dbinom(data.x, ncol(data), 0.25)
      
      data.x <- rescale(data.x, to = c(min(expr.sorted), max(expr.sorted)))
      data.df.dist <- rbind(data.df.dist, data.frame(gene="Binomial distribution (p=0.25)", sample = names(sort(data[ genes, ])), expr=data.x, dens=density2freq(data.y)))
      
      data.x <- 1:ncol(data)
      data.y <- dbinom(data.x, ncol(data), 0.75)
      
      data.x <- rescale(data.x, to = c(min(expr.sorted), max(expr.sorted)))
      data.df.dist <- rbind(data.df.dist, data.frame(gene="Binomial distribution (p=0.75)", sample = names(sort(data[ genes, ])), expr=data.x, dens=density2freq(data.y)))
    }
    
    if ("bimodal" %in% tolower(distributions)){
      
      ##### Draw n/2 samples from a normal distributions with one median and another n/2 samples from a second normal distribution with a different median. Useful link                  https://stats.stackexchange.com/questions/355344/simulating-a-bimodal-distribution-in-the-range-of-15-in-r
      data.x1 <- seq(min(expr.sorted), median(expr.sorted), length.out = ncol(data)/2)
      data.x2 <- seq(median(expr.sorted), max(expr.sorted), length.out = ncol(data)/2)
      
      ##### Combine both normal distributions to generate a bimodal distribution
      data.x <- c(data.x1, data.x2)
      
      ##### Generate y-values for bimodal distribution
      data.y <- c(dnorm(data.x1, mean = mean(data.x1), sd = (max(data.x1)-mean(data.x1))/3), dnorm(data.x2, mean = mean(data.x2), sd = (max(data.x2)-mean(data.x2))/3))
      
      ##### Add bimodal dist values to the distribution dataframe
      data.df.dist <- rbind(data.df.dist, data.frame(gene="Bimodal distribution", sample = names(sort(data[ genes, ])), expr=data.x, dens=density2freq(data.y)))
    }
    
    data.df <- rbind(data.df, data.df.dist)
    
    ##### Extract expression for selected sample in the distributions dataframe
    data.df.selected <- data.frame(matrix(ncol = 4, nrow = 0))
    colnames(data.df.selected) <- c("gene", "sample", "expr", "dens")
    for ( i in 1:length(sampleName) ) {
      data.df.selected <- rbind(data.df.selected, data.df[sampleName[i] == data.df$sample, ])
    }
  }
  
  ##### Get min and max values based on the expression data
  den.x <- sort(data.df$expr)
  den.y <- sort(data.df$dens)
  
  ##### Assign colours to distributions
  genes.colour <- getGenesColours(unique(data.df$gene))
  
  ##### Assign colours to samples
  sample.colour <- getSamplesColours(unique(data.df.selected$sample))

  ##### Generate interactive density plot
  p <- plotly::plot_ly(data.df, x = ~expr, y = ~dens, type = 'scatter', mode = 'lines', color = ~gene, colors = genes.colour, width = 1000, height = 300) %>%
    plotly::add_markers(y = data.df.selected$dens, x = data.df.selected$expr, 
                name = data.df.selected$sample,
                text = data.df.selected$sample,
                mode = 'markers',
                marker = list(size = 8, colors = data.df.selected$sample, color = rep(sample.colour, each = nrow(data.df.selected)/length(samples)), line = list(color = "grey", width = 2)),
                showlegend = TRUE,
                inherit = FALSE) %>%
     plotly::layout(title = main_title,
           xaxis = list(title = x_title, range = c(den.x[1],den.x[length(den.x)])),
           yaxis = list (title = 'Weight', range = c(den.y[1],den.y[length(den.y)])))
  
return( p )
}
```

```{r load_libraries, warning=FALSE}
### Load libraries
suppressMessages(library(preprocessCore))
suppressMessages(library(edgeR))
suppressMessages(library(tidyverse))
suppressMessages(library(package=paste0("EnsDb.Hsapiens.v", params$ensembl_version), character.only = TRUE))
suppressMessages(library(plotly))
suppressMessages(library(scales))
suppressMessages(library(truncnorm))
suppressMessages(library(RColorBrewer))
```

```{r load_data, message=FALSE, warning=FALSE}
##### Read in expression data and associated sample annotation files
data <- read.table(paste(params$exprDir, params$exprFile, sep="/"), sep="\t", as.is=TRUE, header=TRUE, row.names = 1)
targetFile <- read.table(paste(params$exprDir, params$annotFile, sep="/"), sep="\t", as.is=TRUE, header=TRUE)[,c(1:4)]

##### Make sure that there are no duplciated samples in the target file
targetFile <- targetFile[!duplicated(targetFile[,"Sample_name"]),]
rownames(targetFile) <- targetFile[,"Sample_name"]

##### Make syntactically valid names
colnames(data) <- make.names(colnames(data))
rownames(targetFile) <- make.names(rownames(targetFile))

##### Make sure that the target file contains info only about samples present in the data matrix
targetFile <- targetFile[ rownames(targetFile) %in% colnames(data),  ]

#### Prepare input samples of interest
samples <- unique(unlist(strsplit(params$samples, split=',', fixed=TRUE)))

##### Change the target file to mark the sample of interest
targetFile$Target[ rownames(targetFile) %in% samples ] <- "Queried sample(s)"

##### Make sure that the samples order in the data matrix is the same as in the target file. If the expression matrix contains data for additional samples then write the data subset (containing only samples from the target file) into a file
if ( !all( colnames(data) %in% rownames(targetFile) , na.rm = FALSE)  ) {
  
  data <- data[ , rownames(targetFile) ]

  ##### Save data subset into a file
  write.table(geneMatrix2write(data), file=paste0(normalizePath(params$exprDir), "/", paste0(params$exprFile, ".subset.txt")), sep="\t", quote=FALSE, row.names=FALSE, append = FALSE)
}
```

**`r if ( any( samples %!in% colnames(data) ) || any( samples %!in% colnames(data) ) ) { c("NOTE:\n") } else { c(" ") }`** 

`r if ( any( samples %!in% rownames(targetFile) ) ) { c(paste0("\nQueried sample(s) \"", samples[ samples %!in%rownames(targetFile) ], "\" are not present in the \"",  params$annotFile, "\" samples annotation file.\n\n")) }`

`r if ( any( samples %!in% colnames(data) ) ) { c(paste0("\nQueried sample(s) \"", samples[ samples %!in%colnames(data) ], "\" are not present in the \"", params$exprFile, "\" samples annotation file.\n\n")) }`

`r if ( any( samples %!in% colnames(data) ) || any( samples %!in% colnames(data) ) ) { c("These sample(s) will be ignored.\n\n") } else { c(" ") }`

***

## Library size

Bar-plot illustrating library size for each sample.

```{r library_size_plot, message = FALSE, warning=FALSE, fig.width = 12, fig.height = 9}
suppressMessages(library(plotly))
##### Generate bar-plot for library size. The colours indicate sample groups, as provided in *Target* column in the sample annotation file

##### Assigne colours to targets and datasets
targets.colour <- getTargetsColours(targetFile$Target)

##### Prepare data frame
data.df <- data.frame(targetFile$Target, colnames(data), as.numeric(colSums(data)*1e-6))
colnames(data.df) <- c("Group","Sample", "Library_size")

##### The default order will be alphabetized unless specified as below
data.df$Sample <- factor(data.df$Sample, levels = data.df[["Sample"]])
p <- plot_ly(data.df, x = ~Sample, y = ~Library_size, color = ~Group, type = 'bar', colors = targets.colour[[1]], width = 1000, height = 400) %>%
  layout(title = "", xaxis = list( tickfont = list(size = 10), title = ""), yaxis = list(title = "Library size (millions)"), margin = list(l=50, r=50, b=150, t=50, pad=4), autosize = F, legend = list(orientation = 'v', y = 0.5), showlegend=TRUE)

##### Print htmlwidget
p

##### Save the bar-plot as html (PLOTLY)
#htmlwidgets::saveWidget(as_widget(p), paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, "_RNAseq_libSize.html")), selfcontained = TRUE)
##### Detach plotly package. Otherwise it clashes with other graphics devices
detach("package:plotly", unload=FALSE)
```

```{r gene_annot, comment = NA, message=FALSE, warning=FALSE, eval = params$ensembl}
##### Loop through data and annotate genes
##### Convert data into a data frame to make the Ensembl ID and gene symbol matches (with merge function)
data.df <- as.data.frame(cbind(rownames(data), data))
colnames(data.df)[1] <- "ENSEMBL"

##### Get genes annotation and genomic locations
edb <- eval(parse(text = paste0("EnsDb.Hsapiens.v", params$ensembl_version)))
  
##### Get keytypes for gene SYMBOL
keys <- keys(edb, keytype="GENEID")
  
##### Get genes genomic coordiantes
gene_info <- ensembldb::select(edb, keys=keys, columns=c("GENEID", "GENEBIOTYPE", "GENENAME", "SEQNAME", "GENESEQSTART", "GENESEQEND"), keytype="GENEID")
names(gene_info) <- gsub("GENEID", "ENSEMBL", names(gene_info))
names(gene_info) <- gsub("GENENAME", "SYMBOL", names(gene_info))
  
##### Limit genes annotation to those genes for which sample expression measurments are available
gene_info <-  gene_info[ gene_info$ENSEMBL %in% data.df$ENSEMBL,  ]
  
##### Remove rows with duplicated ENSEMBL IDs
gene_info = gene_info[!duplicated(gene_info$ENSEMBL),]
rownames(gene_info) <- gene_info$ENSEMBL
  
##### Remove rows with duplicated gene symbols (Y_RNAs, SNORs, LINC0s etc)
gene_info = gene_info[!duplicated(gene_info$SYMBOL),]
  

##### Merge genes genomic coordinates info with their annotation and expression data
data.annot <- merge(gene_info, data.df, by = "ENSEMBL", all.x = FALSE)
rownames(data.annot) <- data.annot$SYMBOL
  
##### Get data matrix with gene symbols
data <- data.annot[, names(data.annot) %!in% c("SYMBOL", "GENEBIOTYPE", "ENSEMBL", "SEQNAME", "GENESEQSTART", "GENESEQEND")]
```

***

## Data transformation and filtering

The read count data is converted into **`r params$transform`**s using *[edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html){target="_blank"}* functions. `r if ( !params$filter ) { c("The option for filtering out genes with low counts is switched OFF") } else if ( params$filter ) { c("Genes with low counts were filtered out") }`. `r if ( !params$log ) { c("The data is not log-transformed") } else if ( params$log ) { c("The data is log2-transformed") }`. Plot(s) below present `r params$transform` data distribution.

```{r data_transformation_filtering, comment = NA, message=FALSE, warning=FALSE}
##### Filtering to remove low expressed genes. For differential expression and related analyses, gene expression is rarely considered at the level of raw counts since libraries sequenced at a greater depth will result in higher counts. Rather, it is common practice to transform raw counts onto a scale that accounts for such library size differences. Genes with very low counts across all libraries provide little evidence for differential expression. In the biological point of view, a gene must be expressed at some minimal level before it is likely to be translated into a protein or to be biologically important. In addition, the pronounced discretenes of these counts interferes with some of the statistical approximations that are used later in the pipeline. These genes should be filtered out prior to further analysis. Users should filter with CPM rather than filtering on the counts directly, as the latter does not account for differences in library sizes between samples. For instance for the CPM-transformed data we keep only genes that have CPM of 1
##### Transformation to CPM or TPM scale (see these blogs for details https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/ and https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ ).  CPM = Counts Per Million,  TPM = Transcripts Per Kilobase Million. 
##### CPM transformation and filtering
if ( params$filter && params$transform == "CPM" ) {
  
  ##### Create EdgeR DGEList object
  y <- edgeR::DGEList(counts=data,  group=targetFile$Target)
  
  ##### Remove genes with CPM of at least 1 in less than 10% of samples
  #cat("The CPM of 1 (cut-off for removing low expressed genes) corresponds to", round(min(as.numeric(colSums(data)*1e-6)), digits=0), "reads in sample with the lowest sequencing depth, and", round(max(as.numeric(colSums(data)*1e-6)), digits=0), "reads in sample with the greatest sequencing depth\n")
  
  keep <- rowSums(cpm(y)>1) >= ncol(data)/10
  y$filtered <- y[keep, , keep.lib.sizes=FALSE]
  
  #cat(nrow(y$filtered$counts), "genes remained after filtering out of the", nrow(data), "genes in the input read count matrix\n\n")
  
  ##### Transform the raw-scale to CPM. Add small offset to each observation to avoid taking log of zero
  y$samples["norm.factors"] <- edgeR::calcNormFactors(y, method = "none")$samples["norm.factors"]
  y$transformed <- edgeR::cpm(y, normalized.lib.sizes=FALSE, log=params$log, prior.count=0.25)
  y$filtered.transformed <- edgeR::cpm(y$filtered, normalized.lib.sizes=FALSE, log=params$log, prior.count=0.25)
  
##### CPM transformation without filtering
} else if ( !params$filter && params$transform == "CPM" ) {
  
  ##### Create EdgeR DGEList object
  y <- edgeR::DGEList(counts=data,  group=targetFile$Target)
  
  ##### Transform the raw-scale to CPM. Add small offset to each observation to avoid taking log of zero
  y$transformed <- edgeR::cpm(y, normalized.lib.sizes=FALSE, log=params$log, prior.count=0.25)
  
##### TPM data transformation. We can convert RPKM to TPM in two different ways: from pre-calculated RPKM, by diving by the sum of RPKM values, or directly from the normalized counts. Here we calculate TPM starting from RPKM values computed using edgeR's rpkm function ( from http://luisvalesilva.com/datasimple/rna-seq_units.html )
##### TPM transformation with filtering
} else if ( params$filter && params$transform == "TPM" ) {
  
  ##### Get genes lengths
  edb <- eval(parse(text = paste0("EnsDb.Hsapiens.v", params$ensembl_version)))
  
  gene.length <- lengthOf(edb, filter = GeneIdFilter(data.annot$ENSEMBL))
  
  ##### Check for which genes the lenght info is not available and remove them from the data
  genes.no_length <- data.annot$ENSEMBL[ data.annot$ENSEMBL %!in% names(gene.length)]
  data <- data[ data.annot$ENSEMBL %!in% genes.no_length, ]
  
  ##### Create EdgeR DGEList object
  y <- edgeR::DGEList(counts=data,  group=targetFile$Target)
  
  ##### Convert data into RPKM
  y$transformed <- edgeR::rpkm(y, gene.length = gene.length, normalized.lib.sizes=FALSE, log=FALSE)
  
  ##### Remove genes with TPM of at least 0.2 in less than 10% of samples
  keep <- rowSums(y$transformed>0.2) >= ncol(y$transformed)/10
  y$filtered <- y$counts[keep, ]
  y$filtered.transformed <- y$transformed[keep, ]
  
  ##### ... and then to TPM scale. Add small offset to each observation to avoid taking log of zero
  if ( params$log ) {
    
    y$transformed <- log2(tpm_from_rpkm(y$transformed+0.25))
    y$filtered.transformed <- log2(tpm_from_rpkm(y$filtered.transformed+0.25))
  
  } else {
    
    y$transformed <- tpm_from_rpkm(y$transformed)
    y$filtered.transformed <- tpm_from_rpkm(y$filtered.transformed)
  }
  
##### TPM transformation without filtering
} else if ( !params$filter && params$transform == "TPM" ) {
  
  ##### Get genes lengths
  edb <- eval(parse(text = paste0("EnsDb.Hsapiens.v", params$ensembl_version)))
  
  gene.length <- lengthOf(edb, filter = GeneIdFilter(data.annot$ENSEMBL))
  
  ##### Check for which genes the lenght info is not available and remove them from the data
  genes.no_length <- data.annot$ENSEMBL[ data.annot$ENSEMBL %!in% names(gene.length)]
  data <- data[ data.annot$ENSEMBL %!in% genes.no_length, ]
  
  ##### Create EdgeR DGEList object
  y <- edgeR::DGEList(counts=data,  group=targetFile$Target)
  
  ##### Convert data into RPKM
  y$transformed <- edgeR::rpkm(y, gene.length = gene.length, normalized.lib.sizes=FALSE, log=FALSE)
  
  ##### ... and then to TPM scale. Add small offset to each observation to avoid taking log of zero
  if ( params$log ) {
    
    y$transformed <- log2(tpm_from_rpkm(y$transformed+0.25))
  
  } else {
    
    y$transformed <- tpm_from_rpkm(y$transformed)
  }
}
```

```{r data_transformation_plot, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 6}
##### Collect the most extreme density values for set the x-axis and y-axis boundaries
den.x <- density(y$transformed[,1])$x
den.y <- density(y$transformed[,1])$y
  
for (i in 2:ncol(y$transformed)) {
    
  den <- density(y$transformed[,i])
  den.x <- sort(c(den.x, den$x))
  den.y <- sort(c(den.y, den$y))
}
  
if ( params$filter ) {

  par(mfrow=c(1,2))
  
  ##### Before filtering
  plot(density(y$transformed[,1]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=targets.colour[[2]][1])
  title(main="Transformed data (unfiltered)", xlab=params$transform)
  abline(v=0, lty=3)
  
  for (i in 2:ncol(y$transformed)){
    den <- density(y$transformed[,i])
    lines(den$x, den$y, lwd=2, col=targets.colour[[2]][i])
  }
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], bty="n", bg = "transparent")
  
  ##### After filtering
  plot(density(y$filtered.transformed[,1]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=targets.colour[[2]][1])
  title(main="Transformed and filtered data", xlab=params$transform)
  abline(v=0, lty=3)
  
  for (i in 2:ncol(y$filtered.transformed)){
    den <- density(y$filtered.transformed[,i])
    lines(den$x, den$y, lwd=2, col=targets.colour[[2]][i])
  }
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], bty="n", bg = "transparent")
  
  ##### Save the plot as pdf file
  pdf(paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, "_filtering.pdf")), width=8, height=5)
  par(mfrow=c(1,2))
  
  ##### Before filtering
  plot(density(y$transformed[,1]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=targets.colour[[2]][1])
  title(main="Transformed data (unfiltered)", xlab=params$transform)
  abline(v=0, lty=3)
  
  for (i in 2:ncol(y$transformed)){
    den <- density(y$transformed[,i])
    lines(den$x, den$y, lwd=2, col=targets.colour[[2]][i])
  }
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], bty="n", bg = "transparent")
  
  ##### After filtering
  plot(density(y$filtered.transformed[,1]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=targets.colour[[2]][1])
  title(main="Transformed and filtered data", xlab=params$transform)
  abline(v=0, lty=3)
  
  for (i in 2:ncol(y$filtered.transformed)){
    den <- density(y$filtered.transformed[,i])
    lines(den$x, den$y, lwd=2, col=targets.colour[[2]][i])
  }
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], bty="n", bg = "transparent")
  invisible(dev.off())
  
} else {
  
  ##### Without filtering
  plot(density(y$transformed[,1]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=targets.colour[[2]][1])
  title(main="Transformed data (unfiltered)", xlab=params$transform)
  abline(v=0, lty=3)
  
  for (i in 2:ncol(y$transformed)){
    den <- density(y$transformed[,i])
    lines(den$x, den$y, lwd=2, col=targets.colour[[2]][i])
  }
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], bty="n", bg = "transparent")
  
  ##### Save the plot as pdf file
  pdf(paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, "_filtering.pdf")), width=8, height=5)
  ##### Without filtering
  plot(density(y$transformed[,1]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=targets.colour[[2]][1])
  title(main="Transformed data (unfiltered)", xlab=params$transform)
  abline(v=0, lty=3)
  
  for (i in 2:ncol(y$transformed)){
    den <- density(y$transformed[,i])
    lines(den$x, den$y, lwd=2, col=targets.colour[[2]][i])
  }
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], bty="n", bg = "transparent")
  invisible(dev.off())
}
```

***

## Data normalisation

`r if ( tolower(params$norm) == "none" ) { c("No normalisation is performed") } else { c(paste0("During the sample preparation or sequencing process, external factors that are not of biological interest can affect the expression of individual samples. For example, samples processed in the first batch of an experiment can have higher expression overall when compared to samples processed in a second batch. It is assumed that all samples should have a similar range and distribution of expression values. Normalisation for sample-specific effects is required to ensure that the expression distributions of each sample are similar across the entire experiment. Normalisation is performed using ", params$norm, " method")) }`. 

`r if ( tolower(params$norm) != "none" ) { c(paste0("Box-plots below present ", params$transform, " data for individual samples, coloured by sample groups, before and after ", params$norm, " normalisation.")) }`

```{r data_normalisation, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 16}
##### TMM normalsation. Trimmed mean of M-values (https://www.ncbi.nlm.nih.gov/pubmed/20196867) (TMM) is performed using the calcNormFactors function in edgeR. The normalisation factors calculated here are used as a scaling factor for the library sizes. TMM is the recommended for most RNA-Seq data where the majority (more than half) of the genes are believed not differentially expressed between any pair of the samples. It adjusts for RNA composition effect, calculates scaling factors for the library sizes with calcNormFactors function using trimmed mean of M-values (TMM) between each pair of samples. Note, that the raw read counts are used to calculate the normalisation factors
if ( params$transform == "CPM" ) {
  
  ##### Calculate normalization factors and  transformations from the raw-scale to CPM and normalisation using user-defined method
  if ( params$filter ) {
    
    y$noNorm <- y$filtered.transformed
    y$filtered$samples["norm.factors"] <- edgeR::calcNormFactors(y$filtered, method = params$norm)$samples["norm.factors"]
    y$norm <- edgeR::cpm(y$filtered, normalized.lib.sizes=TRUE, log=params$log, prior.count=0.25)
  
  } else {
    
    y$noNorm <- y$transformed
    y$samples["norm.factors"] <- edgeR::calcNormFactors(y, method = params$norm)$samples["norm.factors"]
    y$norm <- edgeR::cpm(y, normalized.lib.sizes=TRUE, log=params$log, prior.count=0.25)
  }
  
##### Quantile normalsation (from https://www.biostars.org/p/296992/ )
} else if ( params$transform == "TPM" ) {
  
  ##### Normalisation using quantile method
  if ( params$filter ) {
    
    y$noNorm <- y$filtered.transformed
    y$filtered.transformed <- data.matrix(y$filtered.transformed) 
    
    if ( tolower(params$norm) != "none" ) {
      
      y$norm  <- normalize.quantiles(y$filtered.transformed, copy = TRUE)
      colnames(y$norm) <- colnames(y$filtered.transformed)
      rownames(y$norm) <- rownames(y$filtered.transformed)
      
    } else {
      
      y$norm  <- y$filtered.transformed
    }
  
  } else {
    
    y$noNorm <- y$transformed
    y$transformed <- data.matrix(y$transformed)
    
    if ( tolower(params$norm) != "none" ) {
      
      y$norm  <- normalize.quantiles(y$transformed, copy = TRUE)
      colnames(y$norm) <- colnames(y$transformed)
      rownames(y$norm) <- rownames(y$transformed)
      
    } else {
      
      y$norm  <- y$transformed
    }
  }
}

##### Set height scaling factor for plots containing normalised data plots
if ( tolower(params$norm) != "none" ) {
  
  norm.height = 2
} else {
  norm.height = 1
}
```

```{r data_normalisation_plots, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 8*norm.height}
##### Plot expression distribution of samples for unnormalised and normalised data
par(mfrow=c(norm.height,1), mar=c((max(nchar(colnames(y$noNorm)))+3)/2, 5, 3, 2))
    
##### Unnormalised data
boxplot(y$noNorm, las=2, col=targets.colour[[2]], main="", pch=".", las=3)
  title(main="Unnormalised data", ylab=params$transform)
legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], horiz=TRUE, bg = "transparent", box.col="transparent")
    
##### Normalised data
if ( tolower(params$norm) != "none" ) {
      
  boxplot(y$norm, las=2, col=targets.colour[[2]], main="", pch=".", las=3)
  title(main=paste0("Normalised data (", params$norm, ")"), ylab=params$transform)
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], horiz=TRUE, bg = "transparent", box.col="transparent")
}
    
##### Save the plot as pdf file
pdf(paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, "_normalisation.pdf")), width=8, height=14)
par(mfrow=c(norm.height,1), mar=c(max(nchar(colnames(y$noNorm)))/2, 5, 3, 2))
    
##### Unnormalised data
boxplot(y$noNorm, las=2, col=targets.colour[[2]], main="", pch=".", las=3)
title(main="Unnormalised data", ylab=params$transform)
    legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], horiz=TRUE, bg = "transparent", box.col="transparent")
      
##### Normalised data
if ( tolower(params$norm) != "none" ) {
      
  boxplot(y$norm, las=2, col=targets.colour[[2]], main="", pch=".", las=3)
  title(main=paste0("Normalised data (", params$norm, ")"), ylab=params$transform)
    legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], horiz=TRUE, bg = "transparent",box.col="transparent")
}
invisible(dev.off())
```

***

## `r if ( tolower(params$scaling) == "sample-wise" ) { c("Z-score transformation") } else { c("Centering") }`

`r if ( tolower(params$scaling) == "sample-wise" ) { c("This step aims to put data from different sources onto the same scale. Z-scores, or standard scores, indicate how many standard deviations an observation is above or below the mean. The z-score reflects a standard normal deviate - the variation of across the standard normal distribution, which is a normal distribution with mean equal to zero and standard deviation equal to one.") } else { c("This step aims to put data from different sources onto the same scale. The centering is done by subtracting the mean expression of all genes in the corresponding sample from the expresion of queried gene in that sample.") }`

Box-plot below presents `r if ( tolower(params$scaling) == "sample-wise" ) { c("z-score transformed") } else { c("centered") }` data (`r if ( tolower(params$scaling) == "sample-wise" ) { c("sample-wise") } else { c("gene-wise") }`) for individual samples, coloured according to corresponding groups.

```{r z_transformation, comment = NA, message=FALSE, warning=FALSE}
##### Perform Z-score transformation/centering
if ( tolower(params$norm) != "none" ) {
    
    if (tolower(params$scaling) == "sample-wise"){
      
      z_scaling <- "sample-wise"
      y$z <- t(scale(t(y$norm)))
    } else {
      
      z_scaling <- "gene-wise"
      y$z <- scale(y$norm, scale = FALSE)
    }
    
} else {
    
    if (tolower(params$scaling) == "sample-wise"){
      
      z_scaling <- "sample-wise"
      y$z <- t(scale(t(y$noNorm)))
    } else {
      
      z_scaling <- "gene-wise"
      y$z <- scale(y$noNorm, scale = FALSE)
    }
}

##### Define axis title for scaled data
if ( tolower(params$scaling) == "sample-wise" ) {
  
  scaled_title <- "Z-score"
} else { 
  scaled_title <-  paste0("Centered ", params$transform)
}
```

```{r data_z_score_plots, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 8}
##### Plot expression distribution of samples for unnormalised and normalised data
par(mfrow=c(1,1), mar=c((max(nchar(colnames(y$z)))+3)/2, 5, 3, 2))
    
boxplot(y$z, las=2, col=targets.colour[[2]], main="", pch=".", las=3)
  title(main=paste0(scaled_title, " data"), ylab=scaled_title)
legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], horiz=TRUE, bg = "transparent", box.col="transparent")
    
##### Save the plot as pdf file
pdf(paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, "_z_scores.pdf")), width=8, height=7)
par(mfrow=c(1,1), mar=c(max(nchar(colnames(y$z)))/2, 5, 3, 2))
    
boxplot(y$z, las=2, col=targets.colour[[2]], main="", pch=".", las=3)
title(main=paste0(scaled_title, " data"), ylab=scaled_title)
    legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], horiz=TRUE, bg = "transparent", box.col="transparent")
invisible(dev.off())
```

***

## Data processing effects {.tabset}

`r if ( tolower(params$norm) != "none" ) { c(paste0("Scatter-plots illustrating the effect of data normalisation and ", if ( tolower(params$scaling) == "sample-wise" ) { c("z-score transformation") } else { c("centering") }, " data (", z_scaling, ") on the ", params$transform, " data. The ", params$transform, " and ", if ( tolower(params$scaling) == "sample-wise" ) { c("z-score data") } else { c(paste0("centered ", params$transform, " values")) }, " , respectively, are plotted as a function of ", paste0("normalised ", params$transform, " (", params$norm, ")"), " values.")) } else { c(paste0("Scatter-plots illustrating the effect of data ", if ( tolower(params$scaling) == "sample-wise" ) { c("z-score transformation") } else { c("centering") }, " (", z_scaling, ") on the ", params$transform, " data. The ", if ( tolower(params$scaling) == "sample-wise" ) { c("z-scores") } else { c(paste0("centered ", params$transform, "s")) }, " are plotted as a function of ", params$transform, " values.")) }`

```{r data_subset, comment = NA, message=FALSE, warning=FALSE}
##### Subset data to include only user-defined genes
genes <- unique(unlist(strsplit(params$genes, split=',', fixed=TRUE)))
y.sub <- y

##### Identify user-defined genes that were not present in the expression matrix
if ( params$filter ) {
    
  genes.missing <- genes[ genes %!in% rownames(y$filtered.transformed) ]
  genes <- genes[ genes %in% rownames(y$filtered.transformed)  ]
    
  keep <- rownames(y$filtered.transformed) %in% genes
  y.sub$filtered.transformed <- y$filtered.transformed[keep, , drop=FALSE]
    
} else {
    
  genes.missing <- genes[ genes %!in% rownames(y$transformed) ]
  genes <- genes[ genes %in% rownames(y$transformed)  ]
    
  keep <- rownames(y$transformed) %in% genes
  y.sub$transformed <- y.sub$transformed[keep, , drop=FALSE]
}

y.sub$noNorm <- y.sub$noNorm[keep, , drop=FALSE]
y.sub$z <- y.sub$z[keep, , drop=FALSE]
  
if ( tolower(params$norm) != "none" ) {
  y.sub$norm <- y.sub$norm[keep, , drop=FALSE]
}

##### Write list of genes into a file
if ( length(genes) > 0 ) {
    write.table(prepare2write(genes), file = paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, ".genes.txt")), sep="\t", quote=FALSE, row.names=TRUE, append = FALSE )
  
  runChunk <- TRUE
} else {
  runChunk <- FALSE
}

##### Write list of missing genes into a file
if ( exists("genes.missing") && length(genes.missing) > 0 ) {
  
  write.table(prepare2write(genes.missing), file = paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, ".genes.missing.txt")), sep="\t", quote=FALSE, row.names=TRUE, append = FALSE )
}
```

**`r if ( !runChunk ) { c("None of the queried genes are not present in the input data.") } else { c("") }`**

### `r if ( !is.na(genes[1]) ) { genes[1] }`

```{r comparison_plot_1, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[1]
plotsNo <- 1

##### Check if queried samples are present in the data matrix and target file. If now, then ignore them
if ( any( samples %!in% rownames(targetFile) ) ) { 
  
  samples <- samples[ samples %in% rownames(targetFile) ]
  
} else if ( any( samples %!in% colnames(data) ) ) { 
  
  samples <- samples[ samples %in% colnames(data) ]
}

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
  
  ##### For normalisation ON
  if ( tolower(params$norm) != "none" ) {
    
    ##### For normalisation ON
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=transformed.data, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=params$transform, sampleName=samples )
      
    ##### Check the effect of data Z-score transformation/centering
    widges.list[[plotsNo+1]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=y.sub$z, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=scaled_title, sampleName=samples )
      
  ##### For normalisation OFF
  } else {
    
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=transformed.data, dataset2=y.sub$z, main_title="", x_title=params$transform, y_title=scaled_title, sampleName=samples )
  }
}

##### Print a list of htmlwidgets
widges.list
```

### `r if ( !is.na(genes[2]) ) { genes[2] }`

```{r comparison_plot_2, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[2]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
  
  ##### For normalisation ON
  if ( tolower(params$norm) != "none" ) {
    
    ##### For normalisation ON
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=transformed.data, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=params$transform, sampleName=samples )
      
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo+1]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=y.sub$z, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=scaled_title, sampleName=samples )
      
  ##### For normalisation OFF
  } else {
    
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=transformed.data, dataset2=y.sub$z, main_title="", x_title=params$transform, y_title=scaled_title, sampleName=samples )
  }
}

##### Print a list of htmlwidgets
widges.list
```

***

### `r if ( !is.na(genes[3]) ) { genes[3] }`

```{r comparison_plot_3, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[3]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
  
  ##### For normalisation ON
  if ( tolower(params$norm) != "none" ) {
    
    ##### For normalisation ON
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=transformed.data, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=params$transform, sampleName=samples )
      
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo+1]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=y.sub$z, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=scaled_title, sampleName=samples )
      
  ##### For normalisation OFF
  } else {
    
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=transformed.data, dataset2=y.sub$z, main_title="", x_title=params$transform, y_title=scaled_title, sampleName=samples )
  }
}

##### Print a list of htmlwidgets
widges.list
```

***

### `r if ( !is.na(genes[4]) ) { genes[4] }`

```{r comparison_plot_4, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[4]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
  
  ##### For normalisation ON
  if ( tolower(params$norm) != "none" ) {
    
    ##### For normalisation ON
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=transformed.data, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=params$transform, sampleName=samples )
      
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo+1]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=y.sub$z, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=scaled_title, sampleName=samples )
      
  ##### For normalisation OFF
  } else {
    
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=transformed.data, dataset2=y.sub$z, main_title="", x_title=params$transform, y_title=scaled_title, sampleName=samples )
  }
}

##### Print a list of htmlwidgets
widges.list
```

***

### `r if ( !is.na(genes[5]) ) { genes[5] }`

```{r comparison_plot_5, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[5]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
  
  ##### For normalisation ON
  if ( tolower(params$norm) != "none" ) {
    
    ##### For normalisation ON
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=transformed.data, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=params$transform, sampleName=samples )
      
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo+1]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=y.sub$z, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=scaled_title, sampleName=samples )
      
  ##### For normalisation OFF
  } else {
    
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=transformed.data, dataset2=y.sub$z, main_title="", x_title=params$transform, y_title=scaled_title, sampleName=samples )
  }
}

##### Print a list of htmlwidgets
widges.list
```

***

### `r if ( !is.na(genes[6]) ) { genes[6] }`

```{r comparison_plot_6, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[6]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
  
  ##### For normalisation ON
  if ( tolower(params$norm) != "none" ) {
    
    ##### For normalisation ON
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=transformed.data, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=params$transform, sampleName=samples )
      
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo+1]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=y.sub$z, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=scaled_title, sampleName=samples )
      
  ##### For normalisation OFF
  } else {
    
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=transformed.data, dataset2=y.sub$z, main_title="", x_title=params$transform, y_title=scaled_title, sampleName=samples )
  }
}

##### Print a list of htmlwidgets
widges.list
```

***

### `r if ( !is.na(genes[7]) ) { genes[7] }`

```{r comparison_plot_7, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[7]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
  
  ##### For normalisation ON
  if ( tolower(params$norm) != "none" ) {
    
    ##### For normalisation ON
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=transformed.data, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=params$transform, sampleName=samples )
      
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo+1]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=y.sub$z, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=scaled_title, sampleName=samples )
      
  ##### For normalisation OFF
  } else {
    
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=transformed.data, dataset2=y.sub$z, main_title="", x_title=params$transform, y_title=scaled_title, sampleName=samples )
  }
}

##### Print a list of htmlwidgets
widges.list
```

***

### `r if ( !is.na(genes[8]) ) { genes[8] }`

```{r comparison_plot_8, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[8]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
  
  ##### For normalisation ON
  if ( tolower(params$norm) != "none" ) {
    
    ##### For normalisation ON
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=transformed.data, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=params$transform, sampleName=samples )
      
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo+1]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=y.sub$z, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=scaled_title, sampleName=samples )
      
  ##### For normalisation OFF
  } else {
    
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=transformed.data, dataset2=y.sub$z, main_title="", x_title=params$transform, y_title=scaled_title, sampleName=samples )
  }
}

##### Print a list of htmlwidgets
widges.list
```

***

### `r if ( !is.na(genes[9]) ) { genes[9] }`

```{r comparison_plot_9, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[9]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
  
  ##### For normalisation ON
  if ( tolower(params$norm) != "none" ) {
    
    ##### For normalisation ON
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=transformed.data, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=params$transform, sampleName=samples )
      
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo+1]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=y.sub$z, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=scaled_title, sampleName=samples )
      
  ##### For normalisation OFF
  } else {
    
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=transformed.data, dataset2=y.sub$z, main_title="", x_title=params$transform, y_title=scaled_title, sampleName=samples )
  }
}

##### Print a list of htmlwidgets
widges.list
```

***

### `r if ( !is.na(genes[10]) ) { genes[10] }`

```{r comparison_plot_10, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[10]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
  
  ##### For normalisation ON
  if ( tolower(params$norm) != "none" ) {
    
    ##### For normalisation ON
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=transformed.data, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=params$transform, sampleName=samples )
      
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo+1]] <- comparePlot(genes=gene, dataset1=y.sub$norm, dataset2=y.sub$z, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), y_title=scaled_title, sampleName=samples )
      
  ##### For normalisation OFF
  } else {
    
    ##### Check the effect of data Z-score transformation / centering
    widges.list[[plotsNo]] <- comparePlot(genes=gene, dataset1=transformed.data, dataset2=y.sub$z, main_title="", x_title=params$transform, y_title=scaled_title, sampleName=samples )
  }
}

##### Print a list of htmlwidgets
widges.list
```

***

## Expression distribution patterns {.tabset}

### Plots {.tabset .tabset-fade}

`r if ( tolower(params$norm) != "none" ) { c(paste0("Plot(s) illustrating ", params$transform, ", normalised ", params$transform, " (", params$norm, ") and ", if ( tolower(params$scaling) == "sample-wise" ) { c("z-score transformed") } else { c("centered") }, " (", z_scaling, ") data distributions for selected gene(s)")) } else { c(paste0("Plot(s) illustrating ", params$transform, " and ", if ( tolower(params$scaling) == "sample-wise" ) { c("z-score transformed") } else { c("centered") }, " (", z_scaling, ") data distributions for selected gene(s)")) }` along with simulated normal, binomial (p=0.25 and p=0.75) and bimodal distributions (computed based on the input data).

**`r if ( !runChunk ) { c("None of the queried genes are not present in the input data.") } else { c("") }`**

#### `r if ( !is.na(genes[1]) ) { genes[1] }`

```{r expr_dist_plot_1, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[1]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
    
  widges.list[[plotsNo]] <- densityPlot(genes=gene, data=transformed.data, main_title="", x_title=params$transform, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
  
  ##### Normalised data 
  if ( tolower(params$norm) != "none" ) {
    
    plotsNo <- plotsNo + 1
    
    #### Interactive gene distribution plot
    widges.list[[plotsNo]] <- densityPlot(genes=gene, data=y.sub$norm, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), sampleName=samples, distributions = c("normal", "binomial", "bimodal") ) 
  }
  widges.list[[plotsNo+1]] <- densityPlot(genes=gene, data=y.sub$z, main_title="", x_title=scaled_title, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
}

##### Print a list of htmlwidgets
widges.list
```

***

#### `r if ( !is.na(genes[2]) ) { genes[2] }`

```{r expr_dist_plot_2, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[2]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
    
  widges.list[[plotsNo]] <- densityPlot(genes=gene, data=transformed.data, main_title="", x_title=params$transform, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
  
  ##### Normalised data 
  if ( tolower(params$norm) != "none" ) {
    
    plotsNo <- plotsNo + 1
    
    #### Interactive gene distribution plot
    widges.list[[plotsNo]] <- densityPlot(genes=gene, data=y.sub$norm, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), sampleName=samples, distributions = c("normal", "binomial", "bimodal") ) 
  }
  widges.list[[plotsNo+1]] <- densityPlot(genes=gene, data=y.sub$z, main_title="", x_title=scaled_title, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
}

##### Print a list of htmlwidgets
widges.list
```

***

#### `r if ( !is.na(genes[3]) ) { genes[3] }`

```{r expr_dist_plot_3, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[3]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
    
  widges.list[[plotsNo]] <- densityPlot(genes=gene, data=transformed.data, main_title="", x_title=params$transform, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
  
  ##### Normalised data 
  if ( tolower(params$norm) != "none" ) {
    
    plotsNo <- plotsNo + 1
    
    #### Interactive gene distribution plot
    widges.list[[plotsNo]] <- densityPlot(genes=gene, data=y.sub$norm, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), sampleName=samples, distributions = c("normal", "binomial", "bimodal") ) 
  }
  widges.list[[plotsNo+1]] <- densityPlot(genes=gene, data=y.sub$z, main_title="", x_title=scaled_title, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
}

##### Print a list of htmlwidgets
widges.list
```

***

#### `r if ( !is.na(genes[4]) ) { genes[4] }`

```{r expr_dist_plot_4, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[4]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
    
  widges.list[[plotsNo]] <- densityPlot(genes=gene, data=transformed.data, main_title="", x_title=params$transform, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
  
  ##### Normalised data 
  if ( tolower(params$norm) != "none" ) {
    
    plotsNo <- plotsNo + 1
    
    #### Interactive gene distribution plot
    widges.list[[plotsNo]] <- densityPlot(genes=gene, data=y.sub$norm, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), sampleName=samples, distributions = c("normal", "binomial", "bimodal") ) 
  }
  widges.list[[plotsNo+1]] <- densityPlot(genes=gene, data=y.sub$z, main_title="", x_title=scaled_title, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
}

##### Print a list of htmlwidgets
widges.list
```

***

#### `r if ( !is.na(genes[5]) ) { genes[5] }`

```{r expr_dist_plot_5, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[5]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
    
  widges.list[[plotsNo]] <- densityPlot(genes=gene, data=transformed.data, main_title="", x_title=params$transform, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
  
  ##### Normalised data 
  if ( tolower(params$norm) != "none" ) {
    
    plotsNo <- plotsNo + 1
    
    #### Interactive gene distribution plot
    widges.list[[plotsNo]] <- densityPlot(genes=gene, data=y.sub$norm, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), sampleName=samples, distributions = c("normal", "binomial", "bimodal") ) 
  }
  widges.list[[plotsNo+1]] <- densityPlot(genes=gene, data=y.sub$z, main_title="", x_title=scaled_title, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
}

##### Print a list of htmlwidgets
widges.list
```

***

#### `r if ( !is.na(genes[6]) ) { genes[6] }`

```{r expr_dist_plot_6, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[6]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
    
  widges.list[[plotsNo]] <- densityPlot(genes=gene, data=transformed.data, main_title="", x_title=params$transform, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
  
  ##### Normalised data 
  if ( tolower(params$norm) != "none" ) {
    
    plotsNo <- plotsNo + 1
    
    #### Interactive gene distribution plot
    widges.list[[plotsNo]] <- densityPlot(genes=gene, data=y.sub$norm, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), sampleName=samples, distributions = c("normal", "binomial", "bimodal") ) 
  }
  widges.list[[plotsNo+1]] <- densityPlot(genes=gene, data=y.sub$z, main_title="", x_title=scaled_title, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
}

##### Print a list of htmlwidgets
widges.list
```

***

#### `r if ( !is.na(genes[7]) ) { genes[7] }`

```{r expr_dist_plot_7, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[7]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
    
  widges.list[[plotsNo]] <- densityPlot(genes=gene, data=transformed.data, main_title="", x_title=params$transform, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
  
  ##### Normalised data 
  if ( tolower(params$norm) != "none" ) {
    
    plotsNo <- plotsNo + 1
    
    #### Interactive gene distribution plot
    widges.list[[plotsNo]] <- densityPlot(genes=gene, data=y.sub$norm, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), sampleName=samples, distributions = c("normal", "binomial", "bimodal") ) 
  }
  widges.list[[plotsNo+1]] <- densityPlot(genes=gene, data=y.sub$z, main_title="", x_title=scaled_title, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
}

##### Print a list of htmlwidgets
widges.list
```

***

#### `r if ( !is.na(genes[8]) ) { genes[8] }`

```{r expr_dist_plot_8, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[8]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
    
  widges.list[[plotsNo]] <- densityPlot(genes=gene, data=transformed.data, main_title="", x_title=params$transform, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
  
  ##### Normalised data 
  if ( tolower(params$norm) != "none" ) {
    
    plotsNo <- plotsNo + 1
    
    #### Interactive gene distribution plot
    widges.list[[plotsNo]] <- densityPlot(genes=gene, data=y.sub$norm, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), sampleName=samples, distributions = c("normal", "binomial", "bimodal") ) 
  }
  widges.list[[plotsNo+1]] <- densityPlot(genes=gene, data=y.sub$z, main_title="", x_title=scaled_title, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
}

##### Print a list of htmlwidgets
widges.list
```

***

#### `r if ( !is.na(genes[9]) ) { genes[9] }`

```{r expr_dist_plot_9, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[9]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
    
  widges.list[[plotsNo]] <- densityPlot(genes=gene, data=transformed.data, main_title="", x_title=params$transform, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
  
  ##### Normalised data 
  if ( tolower(params$norm) != "none" ) {
    
    plotsNo <- plotsNo + 1
    
    #### Interactive gene distribution plot
    widges.list[[plotsNo]] <- densityPlot(genes=gene, data=y.sub$norm, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), sampleName=samples, distributions = c("normal", "binomial", "bimodal") ) 
  }
  widges.list[[plotsNo+1]] <- densityPlot(genes=gene, data=y.sub$z, main_title="", x_title=scaled_title, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
}

##### Print a list of htmlwidgets
widges.list
```

***

#### `r if ( !is.na(genes[10]) ) { genes[10] }`

```{r expr_dist_plot_10, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*(norm.height+1), eval = runChunk}
##### Create a list for htmlwidgets
widges.list <- htmltools::tagList()

gene <- genes[10]
plotsNo <- 1

##### Plot expression distribution in transfromed and normalised data for selected genes
if ( !is.na(gene) ) {
  
  if ( params$filter ) {
    transformed.data <- y.sub$filtered.transformed
  } else {
    transformed.data <- y.sub$transformed
  }
    
  widges.list[[plotsNo]] <- densityPlot(genes=gene, data=transformed.data, main_title="", x_title=params$transform, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
  
  ##### Normalised data 
  if ( tolower(params$norm) != "none" ) {
    
    plotsNo <- plotsNo + 1
    
    #### Interactive gene distribution plot
    widges.list[[plotsNo]] <- densityPlot(genes=gene, data=y.sub$norm, main_title="", x_title=paste0("Normalised ", params$transform, " (", params$norm, ")"), sampleName=samples, distributions = c("normal", "binomial", "bimodal") ) 
  }
  widges.list[[plotsNo+1]] <- densityPlot(genes=gene, data=y.sub$z, main_title="", x_title=scaled_title, sampleName=samples, distributions = c("normal", "binomial",  "bimodal") )
}

##### Print a list of htmlwidgets
widges.list
```

***

### Table

Table with `r params$transform` data values for each sample for selected genes.

**`r if ( runChunk ) { c("Transformed data") } else { c("Queried genes are not present in the input data.") }`**

```{r expr_dist_table_trans, comment = NA, message=FALSE, warning=FALSE, eval = runChunk}
##### Generate a table with expression values for selected genes
if ( params$filter ) {
  
 if ( nrow(y.sub$filtered.transformed) < 11 ) {
    
    DT::datatable( data = round(t(y.sub$filtered.transformed[genes,]), digits = 1), filter="none", rownames = TRUE, extensions = c('Buttons','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, deferRender = TRUE, scrollY = 200, scroller = TRUE), width = 800, caption = htmltools::tags$caption( style = 'caption-side: top; text-align: left; color:grey; font-size:100% ;'), escape = FALSE) %>%
      DT::formatStyle( columns = names(t(y.sub$filtered.transformed)), `font-size` = '12px', 'text-align' = 'center' )
    
  } else {
    cat(paste0(nrow(y.sub$filtered.transformed), " genes were selected. The tables is available for not more than 10 genes!"))
  }

} else {
  
  if ( nrow(y.sub$transformed) < 11 ) {
    
    DT::datatable( data = round(t(y.sub$transformed[genes,]), digits = 1), filter="none", rownames = TRUE, extensions = c('Buttons','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, deferRender = TRUE, scrollY = 200, scroller = TRUE), width = 800, caption = htmltools::tags$caption( style = 'caption-side: top; text-align: left; color:grey; font-size:100% ;'), escape = FALSE) %>%
      DT::formatStyle( columns = names(t(y.sub$transformed)), `font-size` = '12px', 'text-align' = 'center' )
    
  } else {
    cat(paste0(nrow(y.sub$transformed), " genes were selected. The tables is available for not more than 10 genes!"))
  }
}
```

**`r if ( tolower(params$norm) != "none" && runChunk ) { c("Normalised data") }`**

```{r expr_dist_table_normalised, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 6, eval = runChunk}
##### Generate a table with expression values for selected genes
if ( tolower(params$norm) != "none" ) {
  
  if ( nrow(y.sub$norm) < 11 ) {
    
    DT::datatable( data = round(t(y.sub$norm[genes,]), digits = 1), filter="none", rownames = TRUE, extensions = c('Buttons','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, deferRender = TRUE, scrollY = 200, scroller = TRUE), width = 800, caption = htmltools::tags$caption( style = 'caption-side: top; text-align: left; color:grey; font-size:100% ;'), escape = FALSE) %>%
      DT::formatStyle( columns = names(t(y.sub$norm)), `font-size` = '12px', 'text-align' = 'center' )
    
  } else {
    cat(paste0(nrow(y.sub$norm), " genes were selected. The tables is available for not more than 10 genes!"))
  }
}
```

**`r if ( runChunk ) { c("Z-score data") }`**

```{r expr_dist_table_z, comment = NA, message=FALSE, warning=FALSE, eval = runChunk}
##### Generate a table with expression values for selected genes
if ( nrow(y.sub$z) < 11 ) {
    
  DT::datatable( data = round(t(y.sub$z[genes,]), digits = 1), filter="none", rownames = TRUE, extensions = c('Buttons','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, deferRender = TRUE, scrollY = 200, scroller = TRUE), width = 800, caption = htmltools::tags$caption( style = 'caption-side: top; text-align: left; color:grey; font-size:100% ;'), escape = FALSE) %>%
      DT::formatStyle( columns = names(t(y.sub$z)), `font-size` = '12px', 'text-align' = 'center' )
    
  } else {
  cat(paste0(nrow(y.sub$z), " genes were selected. The tables is available for not more than 10 genes!"))
}
```

***

## Addendum

<details>
<summary>Parameters</summary>
<font size="2">

```{r params_info, comment = NA}
for ( i in 1:length(params) ) {
  cat(paste("Parameter: ", names(params)[i], "\nValue: ", paste(unlist(params[i]), collapse = ","), "\n\n", sep=""))
}
```

</font>
</details>

<details>
<summary>Session info</summary>
<font size="2">

```{r sessioninfo, comment = NA}
devtools::session_info()
```

</font>
</details>
