---
title: "Investigate expression data distribution for selected genes"
author: "UMCCR"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
  exprDir: 
  exprFile: 
  annotFile: 
  transform: 
  norm: 
  filter: 
  log: 
  genes: 
  ensembl: 
  results_name: 
  ensembl_version: 86
output:
  html_document:
    keep_md: yes
    code_folding: hide
#    toc: true
#    toc_float: true
---

The aim of this report is investigate the expression data distribution patters of all or user-defined genes in combined expression data (**`r params$exprFile`**) derived from different samples. The data combination includes filtering out genes with low counts `r if ( !params$filter ) { c("(this step is skipped)") }`, transformation (`r params$transform` method, `r if ( !params$log ) { c("with no log-transformation") } else if ( params$log ) { c("followed by log2-transformation") }`) and normalisation (`r params$norm`). The pipeline is based on recommendations from *[RNAseq123](https://master.bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html){target="_blank"}* package.

```{r define_functions, comment=NA, message=FALSE, warning=FALSE}
### Define functions

##### Create 'not in' operator
"%!in%" <- function(x,table) match(x,table, nomatch = 0) == 0
##### Prepare object to write into a file
prepare2write <- function (x) {
  
  x2write <- cbind(rownames(x), x)
  colnames(x2write) <- c("Gene",colnames(x))
  return(x2write)
}
##### Prepare gene data matrix to write into a file
geneMatrix2write <- function (x) {
  
  x2write <- cbind(rownames(x), x)
  colnames(x2write) <- c("Gene",colnames(x))
  return(x2write)
}

##### Assign colours to different groups
getTargetsColours <- function(targets) {
  
  ##### Predefined selection of colours for groups
  targets.colours <- c("red","blue","green","darkgoldenrod","darkred","deepskyblue", "coral", "cornflowerblue", "chartreuse4", "bisque4", "chocolate3", "cadetblue3", "darkslategrey", "lightgoldenrod4", "mediumpurple4", "orangered3","indianred1","blueviolet","darkolivegreen4","darkgoldenrod4","firebrick3","deepskyblue4", "coral3", "dodgerblue1", "chartreuse3", "bisque3", "chocolate4", "cadetblue", "darkslategray4", "lightgoldenrod3", "mediumpurple3", "orangered1")
  
  f.targets <- factor(targets)
  vec.targets <- targets.colours[1:length(levels(f.targets))]
  targets.colour <- rep(0,length(f.targets))
  for(i in 1:length(f.targets))
    targets.colour[i] <- vec.targets[ f.targets[i]==levels(f.targets)]
  
  return( list(vec.targets, targets.colour) )
}

##### Assign colours to different groups
getGenesColours <- function(genes) {
  
  ##### Predefined selection of colours for genes
  genes.colours <- rainbow(length(genes))

  f.genes <- factor(genes)
  vec.genes <- genes.colours[1:length(levels(f.genes))]
  genes.colour <- rep(0,length(f.genes))
  for(i in 1:length(f.genes))
    genes.colour[i] <- vec.genes[ f.genes[i]==levels(f.genes)]
  
  return( genes.colour )
}

##### Calculate TPM from RPKM (from http://luisvalesilva.com/datasimple/rna-seq_units.html )
tpm_from_rpkm <- function(x) {
  rpkm.sum <- colSums(x)
  return(t(t(x) / (1e-06 * rpkm.sum)))
}

##### Generate density plot for selected genes
densityPlot <- function(genes, data, main_title, x_title, sampleName = NULL, distributions = NULL, plot_mode = "static") {
  
  ##### Used data for user-defined genes
  data <- data[ genes, ]
  
  ##### Assigne colours to genes
  genes.colour <- getGenesColours(genes)
  
  ##### Collect the most extreme density values for setting the x-axis and y-axis boundaries
  den.x <- density(data[1,])$x
  den.y <- density(data[1,])$y
  
  for (i in 2:nrow(data)) {
    
    den <- density(data[i,])
    den.x <- sort(c(den.x, den$x))
    den.y <- sort(c(den.y, den$y))
  }
  
  ##### Transformed data for html report
  #### Interactive genes distribution plot
  ##### Create empty data frame
  data.df <- data.frame(matrix(ncol = 3, nrow = 0))
  colnames(data.df) <- c("gene", "expr", "dens")
  
  for (i in 1:nrow(data)) {
    
    data.df <- rbind(data.df, data.frame(gene = genes[i], expr = density(data[i,])$x, dens = density(data[i,])$y))
  }
  
  ##### Generate values to generate various distributions
  if ( !is.null(distributions) ) {
    
    ##### Get min and max values based on the expression data
    data.x <- seq(min(data), max(data), 0.1)
    
    ##### Create empty data frame
    data.df.dist <- data.frame(matrix(ncol = 3, nrow = 0))
    colnames(data.df.dist) <- c("gene", "expr", "dens")
    
    ##### Generate y-values to mirror distributions of interest
    ##### Generate y-values for normal distribution
    if ( "normal" %in% tolower(distributions) ) {
      
      data.y <- dnorm(data.x, mean = mean(data.x), sd = 0.5)
      data.df.dist <- rbind(data.df.dist, data.frame(gene="Normal distribution", expr=data.x, dens=data.y))
      
      den.y <- sort(c(den.y, data.y))
    
      ##### Generate y-values for binomial distribution
    } else if ( "binomial" %in% tolower(distributions) ) {
      
      #data.y.binom <- dbinom(transformed.x, size = 1000, prob = 0.25)
      #data.df.dist <- rbind(data.df.dist, data.frame(gene="Binomial distribution", expr=data.x, dens=data.y))
    
    }
    data.df <- rbind(data.df, data.df.dist)
  }
  
  ##### Generate interactive density plot
  p <- plotly::plot_ly(data.df, x = ~expr, y = ~dens, type = 'scatter', mode = 'lines', color = ~gene, colors = genes.colour, width = 1000, height = 300) %>%
  plotly::layout(title = main_title,
         xaxis = list(title = x_title, range = c(den.x[1],den.x[length(den.x)])),
         yaxis = list (title = 'Density', range = c(den.y[1],den.y[length(den.y)])))
  
    return( p )
}
```

```{r load_libraries, warning=FALSE}
### Load libraries
suppressMessages(library(preprocessCore))
suppressMessages(library(edgeR))
suppressMessages(library(tidyverse))
suppressMessages(library(package=paste0("EnsDb.Hsapiens.v", params$ensembl_version), character.only = TRUE))
suppressMessages(library(plotly))
```

```{r load_data, message=FALSE, warning=FALSE}
##### Read in expression data and associated sample annotation files
data <- read.table(paste(params$exprDir, params$exprFile, sep="/"), sep="\t", as.is=TRUE, header=TRUE, row.names = 1)
targetFile <- read.table(paste(params$exprDir, params$annotFile, sep="/"), sep="\t", as.is=TRUE, header=TRUE)[,c(1:4)]

##### Make sure that there are no duplciated samples in the target file
targetFile <- targetFile[!duplicated(targetFile[,"Sample_name"]),]
rownames(targetFile) <- targetFile[,"Sample_name"]

##### Make syntactically valid names
colnames(data) <- make.names(colnames(data))
rownames(targetFile) <- make.names(rownames(targetFile))

##### Make sure that the target file contains info only about samples present in the data matrix
targetFile <- targetFile[ rownames(targetFile) %in% colnames(data),  ]

##### Make sure that the samples order in the data matrix is the same as in the target file. If the expression matrix contains data for additional samples then write the data subset (containing only samples from the target file) into a file
if ( !all( colnames(data) %in% rownames(targetFile) , na.rm = FALSE)  ) {
  
  data <- data[ , rownames(targetFile) ]

  ##### Save data subset into a file
  write.table(geneMatrix2write(data), file=paste0(normalizePath(params$exprDir), "/", paste0(params$exprFile, ".subset.txt")), sep="\t", quote=FALSE, row.names=FALSE, append = FALSE)
}
```

***

## Library size

Bar-plot illustrating library size for each sample.

```{r library_size_plot, message = FALSE, warning=FALSE, fig.width = 12, fig.height = 9}
suppressMessages(library(plotly))
##### Generate bar-plot for library size. The colours indicate sample groups, as provided in *Target* column in the sample annotation file

##### Assigne colours to targets and datasets
targets.colour <- getTargetsColours(targetFile$Target)

##### Prepare data frame
data.df <- data.frame(targetFile$Target, colnames(data), as.numeric(colSums(data)*1e-6))
colnames(data.df) <- c("Group","Sample", "Library_size")

##### The default order will be alphabetized unless specified as below
data.df$Sample <- factor(data.df$Sample, levels = data.df[["Sample"]])
p <- plot_ly(data.df, x = ~Sample, y = ~Library_size, color = ~Group, type = 'bar', colors = targets.colour[[1]], width = 1000, height = 400) %>%
  layout(title = "", xaxis = list( tickfont = list(size = 10), title = ""), yaxis = list(title = "Library size (millions)"), margin = list(l=50, r=50, b=150, t=50, pad=4), autosize = F, legend = list(orientation = 'v', y = 0.5), showlegend=TRUE)

##### Print htmlwidget
p

##### Save the bar-plot as html (PLOTLY)
#htmlwidgets::saveWidget(as_widget(p), paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, "_RNAseq_libSize.html")), selfcontained = TRUE)
##### Detach plotly package. Otherwise it clashes with other graphics devices
detach("package:plotly", unload=FALSE)
```

```{r gene_annot, comment = NA, message=FALSE, warning=FALSE, eval = params$ensembl}
##### Loop through data and annotate genes
##### Convert data into a data frame to make the Ensembl ID and gene symbol matches (with merge function)
data.df <- as.data.frame(cbind(rownames(data), data))
colnames(data.df)[1] <- "ENSEMBL"

  ##### Get genes annotation and genomic locations
  edb <- eval(parse(text = paste0("EnsDb.Hsapiens.v", params$ensembl_version)))
  
  ##### Get keytypes for gene SYMBOL
  keys <- keys(edb, keytype="GENEID")
  
  ##### Get genes genomic coordiantes
  gene_info <- ensembldb::select(edb, keys=keys, columns=c("GENEID", "GENEBIOTYPE", "GENENAME", "SEQNAME", "GENESEQSTART", "GENESEQEND"), keytype="GENEID")
  names(gene_info) <- gsub("GENEID", "ENSEMBL", names(gene_info))
  names(gene_info) <- gsub("GENENAME", "SYMBOL", names(gene_info))
  
  ##### Limit genes annotation to those genes for which sample expression measurments are available
  gene_info <-  gene_info[ gene_info$ENSEMBL %in% data.df$ENSEMBL,  ]
  
  ##### Remove rows with duplicated ENSEMBL IDs
  gene_info = gene_info[!duplicated(gene_info$ENSEMBL),]
  rownames(gene_info) <- gene_info$ENSEMBL
  
  ##### Remove rows with duplicated gene symbols (Y_RNAs, SNORs, LINC0s etc)
  gene_info = gene_info[!duplicated(gene_info$SYMBOL),]
  
  ##### Add info about immune response markers
  gene_info <- merge(gene_info, ref_genes.list[["genes_immune"]], by = "SYMBOL", all.x = TRUE)
  
  ##### Keep only immune response marjers for which there is available annotation
  ref_genes.list[["genes_immune"]] <- ref_genes.list[["genes_immune"]][ ref_genes.list[["genes_immune"]]$SYMBOL %in% gene_info$SYMBOL, ]
  
  ##### Merge genes genomic coordinates info with their annotation and expression data
  data.annot <- merge(gene_info, data.df, by = "ENSEMBL", all.x = FALSE)
  rownames(data.annot) <- data.annot$ENSEMBL
  
  ##### Get data matrix with gene symbols
  ref_datasets.list[[tissue]][["gene_annot_all"]] <- data.annot[, c("SYMBOL", "GENEBIOTYPE", "ENSEMBL", "SEQNAME", "GENESEQSTART", "GENESEQEND", "Immune_Cycle_Role")]
```

## Data transformation and filtering

The read count data is converted into **`r params$transform`**s using *[edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html){target="_blank"}* functions. `r if ( !params$filter ) { c("The option for filtering out genes with low counts is switched OFF") } else if ( params$filter ) { c("Genes with low counts were filtered out") }`. `r if ( !params$log ) { c("The data is not log-transformed") } else if ( params$log ) { c("The data is log2-transformed") }`. Plot(s) below present `r params$transform` data distribution.

```{r data_transformation_filtering, comment = NA, message=FALSE, warning=FALSE}
##### Filtering to remove low expressed genes. For differential expression and related analyses, gene expression is rarely considered at the level of raw counts since libraries sequenced at a greater depth will result in higher counts. Rather, it is common practice to transform raw counts onto a scale that accounts for such library size differences. Genes with very low counts across all libraries provide little evidence for differential expression. In the biological point of view, a gene must be expressed at some minimal level before it is likely to be translated into a protein or to be biologically important. In addition, the pronounced discretenes of these counts interferes with some of the statistical approximations that are used later in the pipeline. These genes should be filtered out prior to further analysis. Users should filter with CPM rather than filtering on the counts directly, as the latter does not account for differences in library sizes between samples. For instance for the CPM-transformed data we keep only genes that have CPM of 1
##### Transformation to CPM or TPM scale (see these blogs for details https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/ and https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ ).  CPM = Counts Per Million,  TPM = Transcripts Per Kilobase Million. 
##### CPM transformation and filtering
if ( params$filter && params$transform == "CPM" ) {
  
  ##### Create EdgeR DGEList object
  y <- edgeR::DGEList(counts=data,  group=targetFile$Target)
  
  ##### Remove genes with CPM of at least 1 in less than 10% of samples
  #cat("The CPM of 1 (cut-off for removing low expressed genes) corresponds to", round(min(as.numeric(colSums(data)*1e-6)), digits=0), "reads in sample with the lowest sequencing depth, and", round(max(as.numeric(colSums(data)*1e-6)), digits=0), "reads in sample with the greatest sequencing depth\n")
  
  keep <- rowSums(cpm(y)>1) >= ncol(data)/10
  y$filtered <- y[keep, , keep.lib.sizes=FALSE]
  
  #cat(nrow(y$filtered$counts), "genes remained after filtering out of the", nrow(data), "genes in the input read count matrix\n\n")
  
  ##### Transform the raw-scale to CPM. Add small offset to each observation to avoid taking log of zero
  y$samples["norm.factors"] <- edgeR::calcNormFactors(y, method = "none")$samples["norm.factors"]
  y$transformed <- edgeR::cpm(y, normalized.lib.sizes=FALSE, log=params$log, prior.count=0.25)
  y$filtered.transformed <- edgeR::cpm(y$filtered, normalized.lib.sizes=FALSE, log=params$log, prior.count=0.25)
  
##### CPM transformation without filtering
} else if ( !params$filter && params$transform == "CPM" ) {
  
  ##### Create EdgeR DGEList object
  y <- edgeR::DGEList(counts=data,  group=targetFile$Target)
  
  ##### Transform the raw-scale to CPM. Add small offset to each observation to avoid taking log of zero
  y$transformed <- edgeR::cpm(y, normalized.lib.sizes=FALSE, log=params$log, prior.count=0.25)
  
##### TPM data transformation. We can convert RPKM to TPM in two different ways: from pre-calculated RPKM, by diving by the sum of RPKM values, or directly from the normalized counts. Here we calculate TPM starting from RPKM values computed using edgeR's rpkm function ( from http://luisvalesilva.com/datasimple/rna-seq_units.html )
##### TPM transformation with filtering
} else if ( params$filter && params$transform == "TPM" ) {
  
  ##### Get genes lengths
  edb <- eval(parse(text = paste0("EnsDb.Hsapiens.v", params$ensembl_version)))
  
  gene.length <- lengthOf(edb, filter = GeneIdFilter(rownames(data)))
  
  ##### Check for which genes the lenght info is not available and remove them from the data
  genes.no_length <- rownames(data)[ rownames(data) %!in% names(gene.length)]
  data <- data[ rownames(data) %!in% genes.no_length, ]
  
  ##### Create EdgeR DGEList object
  y <- edgeR::DGEList(counts=data,  group=targetFile$Target)
  
  ##### Convert data into RPKM
  y$transformed <- edgeR::rpkm(y, gene.length = gene.length, normalized.lib.sizes=FALSE, log=FALSE)
  
  ##### Remove genes with TPM of at least 0.2 in less than 10% of samples
  keep <- rowSums(y$transformed>0.2) >= ncol(y$transformed)/10
  y$filtered <- y$counts[keep, ]
  y$filtered.transformed <- y$transformed[keep, ]
  
  ##### ... and then to TPM scale. Add small offset to each observation to avoid taking log of zero
  if ( params$log ) {
    
    y$transformed <- log2(tpm_from_rpkm(y$transformed+0.25))
    y$filtered.transformed <- log2(tpm_from_rpkm(y$filtered.transformed+0.25))
  
  } else {
    
    y$transformed <- tpm_from_rpkm(y$transformed)
    y$filtered.transformed <- tpm_from_rpkm(y$filtered.transformed)
  }
  
##### TPM transformation without filtering
} else if ( !params$filter && params$transform == "TPM" ) {
  
  ##### Get genes lengths
  edb <- eval(parse(text = paste0("EnsDb.Hsapiens.v", params$ensembl_version)))
  
  gene.length <- lengthOf(edb, filter = GeneIdFilter(rownames(data)))
  
  ##### Check for which genes the lenght info is not available and remove them from the data
  genes.no_length <- rownames(data)[ rownames(data) %!in% names(gene.length)]
  data <- data[ rownames(data) %!in% genes.no_length, ]
  
  ##### Create EdgeR DGEList object
  y <- edgeR::DGEList(counts=data,  group=targetFile$Target)
  
  ##### Convert data into RPKM
  y$transformed <- edgeR::rpkm(y, gene.length = gene.length, normalized.lib.sizes=FALSE, log=FALSE)
  
  ##### ... and then to TPM scale. Add small offset to each observation to avoid taking log of zero
  if ( params$log ) {
    
    y$transformed <- log2(tpm_from_rpkm(y$transformed+0.25))
  
  } else {
    
    y$transformed <- tpm_from_rpkm(y$transformed)
  }
}
```

```{r data_transformation_plot, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 6}
##### Collect the most extreme density values for set the x-axis and y-axis boundaries
den.x <- density(y$transformed[,1])$x
den.y <- density(y$transformed[,1])$y
  
for (i in 2:ncol(y$transformed)) {
    
  den <- density(y$transformed[,i])
  den.x <- sort(c(den.x, den$x))
  den.y <- sort(c(den.y, den$y))
}
  
if ( params$filter ) {

  par(mfrow=c(1,2))
  
  ##### Before filtering
  plot(density(y$transformed[,1]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=targets.colour[[2]][1])
  title(main="Transformed data (unfiltered)", xlab=params$transform)
  abline(v=0, lty=3)
  
  for (i in 2:ncol(y$transformed)){
    den <- density(y$transformed[,i])
    lines(den$x, den$y, lwd=2, col=targets.colour[[2]][i])
  }
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], bty="n", bg = "transparent")
  
  ##### After filtering
  plot(density(y$filtered.transformed[,1]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=targets.colour[[2]][1])
  title(main="Transformed and filtered data", xlab=params$transform)
  abline(v=0, lty=3)
  
  for (i in 2:ncol(y$filtered.transformed)){
    den <- density(y$filtered.transformed[,i])
    lines(den$x, den$y, lwd=2, col=targets.colour[[2]][i])
  }
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], bty="n", bg = "transparent")
  
  ##### Save the plot as pdf file
  pdf(paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, "_filtering.pdf")), width=8, height=5)
  par(mfrow=c(1,2))
  
  ##### Before filtering
  plot(density(y$transformed[,1]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=targets.colour[[2]][1])
  title(main="Transformed data (unfiltered)", xlab=params$transform)
  abline(v=0, lty=3)
  
  for (i in 2:ncol(y$transformed)){
    den <- density(y$transformed[,i])
    lines(den$x, den$y, lwd=2, col=targets.colour[[2]][i])
  }
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], bty="n", bg = "transparent")
  
  ##### After filtering
  plot(density(y$filtered.transformed[,1]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=targets.colour[[2]][1])
  title(main="Transformed and filtered data", xlab=params$transform)
  abline(v=0, lty=3)
  
  for (i in 2:ncol(y$filtered.transformed)){
    den <- density(y$filtered.transformed[,i])
    lines(den$x, den$y, lwd=2, col=targets.colour[[2]][i])
  }
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], bty="n", bg = "transparent")
  invisible(dev.off())
  
} else if ( !params$filter ) {
  
  ##### Without filtering
  plot(density(y$transformed[,1]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=targets.colour[[2]][1])
  title(main="Transformed data (unfiltered)", xlab=params$transform)
  abline(v=0, lty=3)
  
  for (i in 2:ncol(y$transformed)){
    den <- density(y$transformed[,i])
    lines(den$x, den$y, lwd=2, col=targets.colour[[2]][i])
  }
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], bty="n", bg = "transparent")
  
  ##### Save the plot as pdf file
  pdf(paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, "_filtering.pdf")), width=8, height=5)
  ##### Without filtering
  plot(density(y$transformed[,1]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=targets.colour[[2]][1])
  title(main="Transformed data (unfiltered)", xlab=params$transform)
  abline(v=0, lty=3)
  
  for (i in 2:ncol(y$transformed)){
    den <- density(y$transformed[,i])
    lines(den$x, den$y, lwd=2, col=targets.colour[[2]][i])
  }
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], bty="n", bg = "transparent")
  invisible(dev.off())
}
```

## Data normalisation

`r if ( tolower(params$norm) == "none" ) { c("No normalisation is performed") } else { c(paste0("During the sample preparation or sequencing process, external factors that are not of biological interest can affect the expression of individual samples. For example, samples processed in the first batch of an experiment can have higher expression overall when compared to samples processed in a second batch. It is assumed that all samples should have a similar range and distribution of expression values. Normalisation for sample-specific effects is required to ensure that the expression distributions of each sample are similar across the entire experiment. Normalisation is performed using ", params$norm, " method")) }`. 

`r if ( tolower(params$norm) != "none" ) { c(paste0("Box-plots below present ", params$transform, " data for individual samples, coloured by sample groups, before and after ", params$norm, " normalisation.")) }`

```{r data_normalisation, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 16}
##### TMM normalsation. Trimmed mean of M-values (https://www.ncbi.nlm.nih.gov/pubmed/20196867) (TMM) is performed using the calcNormFactors function in edgeR. The normalisation factors calculated here are used as a scaling factor for the library sizes. TMM is the recommended for most RNA-Seq data where the majority (more than half) of the genes are believed not differentially expressed between any pair of the samples. It adjusts for RNA composition effect, calculates scaling factors for the library sizes with calcNormFactors function using trimmed mean of M-values (TMM) between each pair of samples. Note, that the raw read counts are used to calculate the normalisation factors
if ( params$transform == "CPM" ) {
  
  ##### Calculate normalization factors and  transformations from the raw-scale to CPM and normalisation using user-defined method
  if ( params$filter ) {
    
    y$noNorm <- y$filtered.transformed
    y$filtered$samples["norm.factors"] <- edgeR::calcNormFactors(y$filtered, method = params$norm)$samples["norm.factors"]
    y$norm <- edgeR::cpm(y$filtered, normalized.lib.sizes=TRUE, log=params$log, prior.count=0.25)
  
  } else {
    
    y$noNorm <- y$transformed
    y$samples["norm.factors"] <- edgeR::calcNormFactors(y, method = params$norm)$samples["norm.factors"]
    y$norm <- edgeR::cpm(y, normalized.lib.sizes=TRUE, log=params$log, prior.count=0.25)
  }
  
##### Quantile normalsation (from https://www.biostars.org/p/296992/ )
} else if ( params$transform == "TPM" ) {
  
  ##### Normalisation using quantile method
  if ( params$filter ) {
    
    y$noNorm <- y$filtered.transformed
    y$filtered.transformed <- data.matrix(y$filtered.transformed) 
    
    if ( tolower(params$norm) != "none" ) {
      
      y$norm  <- normalize.quantiles(y$filtered.transformed, copy = TRUE)
      colnames(y$norm) <- colnames(y$filtered.transformed)
      rownames(y$norm) <- rownames(y$filtered.transformed)
      
    } else {
      
      y$norm  <- y$filtered.transformed
    }
  
  } else {
    
    y$noNorm <- y$transformed
    y$transformed <- data.matrix(y$transformed)
    
    if ( tolower(params$norm) != "none" ) {
      
      y$norm  <- normalize.quantiles(y$transformed, copy = TRUE)
      colnames(y$norm) <- colnames(y$transformed)
      rownames(y$norm) <- rownames(y$transformed)
      
    } else {
      
      y$norm  <- y$transformed
    }
  }
}

##### Set height scaling factor for plots containing normalised data plots
if ( tolower(params$norm) != "none" ) {
  
  norm.height = 2
} else {
  norm.height = 1
}
```

```{r data_normalisation_plots, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 8*norm.height}
##### Plot expression distribution of samples for unnormalised and normalised data
par(mfrow=c(norm.height,1), mar=c((max(nchar(colnames(y$noNorm)))+3)/2, 5, 3, 2))
    
##### Unnormalised data
boxplot(y$noNorm, las=2, col=targets.colour[[2]], main="", pch=".", las=3)
  title(main="Unnormalised data", ylab=params$transform)
legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], horiz=TRUE, bg = "transparent", box.col="transparent")
    
##### Normalised data
if ( tolower(params$norm) != "none" ) {
      
  boxplot(y$norm, las=2, col=targets.colour[[2]], main="", pch=".", las=3)
  title(main=paste0("Normalised data (", params$norm, ")"), ylab=params$transform)
  legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], horiz=TRUE, bg = "transparent", box.col="transparent")
}
    
##### Save the plot as pdf file
pdf(paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, "_normalisation.pdf")), width=8, height=14)
par(mfrow=c(norm.height,1), mar=c(max(nchar(colnames(y$noNorm)))/2, 5, 3, 2))
    
##### Unnormalised data
boxplot(y$noNorm, las=2, col=targets.colour[[2]], main="", pch=".", las=3)
title(main="Unnormalised data", ylab=params$transform)
    legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], horiz=TRUE, bg = "transparent", box.col="transparent")
      
##### Normalised data
if ( tolower(params$norm) != "none" ) {
      
  boxplot(y$norm, las=2, col=targets.colour[[2]], main="", pch=".", las=3)
  title(main=paste0("Normalised data (", params$norm, ")"), ylab=params$transform)
    legend("topright", legend=levels(factor(targetFile$Target)), fill=targets.colour[[1]], horiz=TRUE, bg = "transparent",box.col="transparent")
}
invisible(dev.off())
```

```{r z_transformation, comment = NA, message=FALSE, warning=FALSE}
##### Perform Z-score transformation
if ( tolower(params$norm) == "none" ) {
  
    y$z <- scale(y$norm)
    
} else {
    y$z <- scale(y$noNorm)
}
```

## Expression distribution patterns {.tabset}

```{r data_subset, comment = NA, message=FALSE, warning=FALSE}
##### Subset data to include only user-defined genes
if ( tolower(params$genes) != "all"  ) {
    
  genes <- unique(unlist(strsplit(params$genes, split=',', fixed=TRUE)))
  
  ##### Identify user-defined genes that were not present in the expression matrix
  if ( params$filter ) {
    
    genes.missing <- genes[ genes %!in% rownames(y$filtered.transformed) ]

    genes <- genes[ genes %in% rownames(y$filtered.transformed)  ]
    
    keep <- rownames(y$filtered.transformed) %in% genes
    y.sub <- y[keep, , keep.lib.sizes=FALSE]
    y.sub$filtered.transformed <- y.sub$filtered.transformed[keep, ]
  
  } else {
    
    genes.missing <- genes[ genes %!in% rownames(y$transformed) ]

    genes <- genes[ genes %in% rownames(y$transformed)  ]
    
    keep <- rownames(y$transformed) %in% genes
    y.sub <- y[keep, , keep.lib.sizes=FALSE]
    y.sub$transformed <- y.sub$transformed[keep, ]
  }
  
  y.sub$noNorm <- y.sub$noNorm[keep, ]
  y.sub$z <- y.sub$z[keep, ]
  
  if ( tolower(params$norm) != "none" ) {
    y.sub$norm <- y.sub$norm[keep, ]
  }
  
} else {
  
  y.sub <- y
  genes <- rownames(data)
}

##### Write list of genes into a file
if ( length(genes) > 0 ) {
    write.table(prepare2write(genes), file = paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, ".genes.txt")), sep="\t", quote=FALSE, row.names=TRUE, append = FALSE )
  
  runChunk <- TRUE
} else {
  runChunk <- FALSE
}

##### Write list of missing genes into a file
if ( exists("genes.missing") && length(genes.missing) > 0 ) {
  
  write.table(prepare2write(genes.missing), file = paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, ".genes.missing.txt")), sep="\t", quote=FALSE, row.names=TRUE, append = FALSE )
}
```

### Plot

Plot presenting `r params$transform` data distribution for selected gene(s).

```{r expr_dist_plot, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 5*norm.height, eval = runChunk}
##### Plot expression distribution in transfromed and normalised data for selected genes
if ( params$filter ) {
  
  densityPlot(genes=genes, data=y.sub$filtered.transformed, main_title="Transformed data", x_title=params$transform, distributions = "normal") 
} else {
    densityPlot(genes=genes, data=y.sub$transformed, main_title="Transformed data", x_title=params$transform, distributions = "normal") 
}

##### Normalised data
if ( tolower(params$norm) != "none" ) {
  
  #### Interactive genes distribution plot
  densityPlot(genes=genes, data=y.sub$norm, main_title="Normalised data", x_title=params$transform, distributions = "normal") 
}

densityPlot(genes=genes, data=y.sub$z, main_title="Z-scores data", x_title=params$transform, distributions = "normal") 

##### Save the plot as pdf file
genes.colour <- getGenesColours(genes)

pdf(paste0(normalizePath(params$exprDir), "/", paste0(params$results_name, "_expr_distributions.pdf")), width=8, height=5*norm.height)
par(mfrow=c(norm.height,1))

##### Transformed data for PDF
plot(density(y.sub$transformed[1,]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=genes.colour[1])
title(main="Transformed data", xlab=params$transform)
abline(v=0, lty=3)
  
for (i in 2:nrow(y.sub$transformed)){
  den <- density(y.sub$transformed[i,])
  lines(den$x, den$y, lwd=2, col=genes.colour[i])
}
legend("topright", legend=rownames(y.sub$transformed), fill=genes.colour, bty="n", bg = "transparent")
  
##### Normalised data
if ( tolower(params$norm) != "none" ) {
  
  ##### Normalised data
  plot(density(y.sub$norm[1,]), lwd=2, xlim=c(den.x[1],den.x[length(den.x)]), ylim=c(den.y[1],den.y[length(den.y)]), las=2, main="", xlab="", col=genes.colour[1])
  title(main="Normalised data", xlab=params$transform)
  abline(v=0, lty=3)
      
  for (i in 2:nrow(y.sub$norm)) {
    den <- density(y.sub$norm[i,])
    lines(den$x, den$y, lwd=2, col=genes.colour[i])
  }
  legend("topright", legend=rownames(y.sub$norm), fill=genes.colour, bty="n", bg = "transparent")
} 
invisible(dev.off())
```

### Table

Table with `r params$transform` data values for each sample for selected genes.

**Transformed data**

```{r expr_dist_table_trans, comment = NA, message=FALSE, warning=FALSE, eval = runChunk}
##### Generate a table with expression values for selected genes
if ( nrow(y.sub$transformed) < 25 ) {
  
  DT::datatable( data = round(t(y.sub$transformed), digits = 1), filter="none", rownames = TRUE, extensions = c('Buttons','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, deferRender = TRUE, scrollY = 200, scroller = TRUE), width = 800, caption = htmltools::tags$caption( style = 'caption-side: top; text-align: left; color:grey; font-size:100% ;'), escape = FALSE) %>%
    DT::formatStyle( columns = names(t(y.sub$transformed)), `font-size` = '12px', 'text-align' = 'center' )
  
} else {
  cat(paste0(nrow(y.sub$transformed), " genes were selected. The tables is available only if less than 25 genes are selected!"))
}
```

**`r if ( tolower(params$norm) != "none" ) { c("Normalised data") }`**

```{r expr_dist_table_normalised, comment = NA, message=FALSE, warning=FALSE, fig.width = 12, fig.height = 6, eval = runChunk}
##### Generate a table with expression values for selected genes
if ( tolower(params$norm) != "none" ) {
  
  if ( nrow(y.sub$norm) < 25 ) {
    
    DT::datatable( data = round(t(y.sub$norm), digits = 1), filter="none", rownames = TRUE, extensions = c('Buttons','Scroller'), options = list(pageLength = 10, dom = 'Bfrtip', buttons = c('excel', 'csv', 'pdf','copy','colvis'), scrollX = TRUE, deferRender = TRUE, scrollY = 200, scroller = TRUE), width = 800, caption = htmltools::tags$caption( style = 'caption-side: top; text-align: left; color:grey; font-size:100% ;'), escape = FALSE) %>%
      DT::formatStyle( columns = names(t(y.sub$norm)), `font-size` = '12px', 'text-align' = 'center' )
    
  } else {
    cat(paste0(nrow(y.sub$norm), " genes were selected. The tables is available only if less than 25 genes are selected!"))
  }
}
```


***

## Addendum

<details>
<summary>Parameters</summary>
<font size="2">

```{r params_info, comment = NA}
for ( i in 1:length(params) ) {
  cat(paste("Parameter: ", names(params)[i], "\nValue: ", paste(unlist(params[i]), collapse = ","), "\n\n", sep=""))
}
```

</font>
</details>

<details>
<summary>Session info</summary>
<font size="2">

```{r sessioninfo, comment = NA}
devtools::session_info()
```

</font>
</details>
